{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import snorkel\n",
    "from snorkel.labeling import labeling_function\n",
    "from snorkel.labeling import LabelingFunction\n",
    "from snorkel.labeling import PandasLFApplier\n",
    "from snorkel.labeling import LFAnalysis\n",
    "import re\n",
    "from snorkel.labeling.model import MajorityLabelVoter\n",
    "import json\n",
    "import pandas as pd\n",
    "import label_improve"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import label_improve as li"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading the data \n",
    "dataset_name = \"chemprot\"\n",
    "\n",
    "idx_to_label = json.load(open(f\"../weak_datasets/{dataset_name}/label.json\"))\n",
    "label_to_idx = {l:i for i,l in idx_to_label.items()}\n",
    "valid_df = li.chemprot_to_df(json.load(open(f\"../weak_datasets/{dataset_name}/valid.json\", \"r\")))\n",
    "train_df = li.chemprot_to_df(json.load(open(f\"../weak_datasets/{dataset_name}/train.json\", \"r\")))\n",
    "test_df = li.chemprot_to_df(json.load(open(f\"../weak_datasets/{dataset_name}/test.json\", \"r\")))\n",
    "\n",
    "# Sample a dev set to help seed ideas for LFs\n",
    "dev_df = train_df.sample(250, random_state=123)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# chemprot functions:\n",
    "\n",
    "ABSTAIN = -1\n",
    "### Keyword based labeling functions ###\n",
    "\n",
    "## Part of\n",
    "#0\n",
    "@labeling_function()\n",
    "def lf_amino_acid(x):\n",
    "    return 0 if 'amino acid' in x.text.lower() else ABSTAIN\n",
    "#1\n",
    "@labeling_function()\n",
    "def lf_replace(x):\n",
    "    return 0 if 'replace' in x.text.lower() else ABSTAIN\n",
    "#2 TODO: 0.1988\n",
    "@labeling_function()\n",
    "def lf_mutant(x):\n",
    "    return 0 if 'mutant' in x.text.lower() or 'mutat' in x.text.lower() else ABSTAIN\n",
    "#3\n",
    "## Regulator\n",
    "@labeling_function()\n",
    "def lf_bind(x):\n",
    "    return 1 if 'bind' in x.text.lower() else ABSTAIN\n",
    "#4\n",
    "@labeling_function()\n",
    "def lf_interact(x):\n",
    "    return 1 if 'interact' in x.text.lower() else ABSTAIN\n",
    "#5\n",
    "@labeling_function()\n",
    "def lf_affinity(x):\n",
    "    return 1 if 'affinit' in x.text.lower() else ABSTAIN\n",
    "#6 TODO: 0.3578\n",
    "## Upregulator\n",
    "# Activator\n",
    "@labeling_function()\n",
    "def lf_activate(x):\n",
    "    return 2 if 'activat' in x.text.lower() else ABSTAIN\n",
    "#7\n",
    "@labeling_function()\n",
    "def lf_increase(x):\n",
    "    return 2 if 'increas' in x.text.lower() else ABSTAIN\n",
    "#8 TODO: \n",
    "@labeling_function()\n",
    "def lf_induce(x):\n",
    "    return 2 if 'induc' in x.text.lower() else ABSTAIN\n",
    "#9 TODO: \n",
    "@labeling_function()\n",
    "def lf_stimulate(x):\n",
    "    return 2 if 'stimulat' in x.text.lower() else ABSTAIN\n",
    "#10\n",
    "@labeling_function()\n",
    "def lf_upregulate(x):\n",
    "    return 2 if 'upregulat' in x.text.lower() else ABSTAIN\n",
    "#11\n",
    "## Downregulator\n",
    "@labeling_function()\n",
    "def lf_downregulate(x):\n",
    "    return 3 if 'downregulat' in x.text.lower() or 'down-regulat' in x.text.lower() else ABSTAIN\n",
    "#12\n",
    "@labeling_function()\n",
    "def lf_reduce(x):\n",
    "    return 3 if 'reduc' in x.text.lower() else ABSTAIN\n",
    "#13\n",
    "@labeling_function()\n",
    "def lf_inhibit(x):\n",
    "    return 3 if 'inhibit' in x.text.lower() else ABSTAIN\n",
    "#14\n",
    "@labeling_function()\n",
    "def lf_decrease(x):\n",
    "    return 3 if 'decreas' in x.text.lower() else ABSTAIN\n",
    "#15\n",
    "## Agonist\n",
    "@labeling_function()\n",
    "def lf_agonist(x):\n",
    "    return 4 if ' agoni' in x.text.lower() or \"\\tagoni\" in x.text.lower() else ABSTAIN\n",
    "\n",
    "#16\n",
    "## Antagonist\n",
    "@labeling_function()\n",
    "def lf_antagonist(x):\n",
    "    return 5 if 'antagon' in x.text.lower() else ABSTAIN\n",
    "\n",
    "#17\n",
    "## Modulator\n",
    "# TODO: Delete this LF, or change this to modulator ??\n",
    "@labeling_function()\n",
    "def lf_modulate(x):\n",
    "    return 6 if 'modulat' in x.text.lower() else ABSTAIN\n",
    "\n",
    "#18\n",
    "@labeling_function()\n",
    "def lf_allosteric(x):\n",
    "    return 6 if 'allosteric' in x.text.lower() else ABSTAIN\n",
    "#19\n",
    "## Cofactor\n",
    "@labeling_function()\n",
    "def lf_cofactor(x):\n",
    "    return 7 if 'cofactor' in x.text.lower() else ABSTAIN\n",
    "#20\n",
    "## Substrate/Product\n",
    "@labeling_function()\n",
    "def lf_substrate(x):\n",
    "    return 8 if 'substrate' in x.text.lower() else ABSTAIN\n",
    "#21\n",
    "@labeling_function()\n",
    "def lf_transport(x):\n",
    "    return 8 if 'transport' in x.text.lower() else ABSTAIN\n",
    "#22\n",
    "@labeling_function()\n",
    "def lf_catalyze(x):\n",
    "    return 8 if 'catalyz' in x.text.lower() or 'catalys' in x.text.lower() else ABSTAIN\n",
    "#23\n",
    "@labeling_function()\n",
    "def lf_product(x):\n",
    "    return 8 if \"produc\" in x.text.lower() else ABSTAIN\n",
    "#24\n",
    "@labeling_function()\n",
    "def lf_convert(x):\n",
    "    return 8 if \"conver\" in x.text.lower() else ABSTAIN\n",
    "#25\n",
    "## NOT\n",
    "@labeling_function()\n",
    "def lf_not(x):\n",
    "    return 9 if 'not' in x.text.lower() else ABSTAIN\n",
    "\n",
    "# 26 replace the 17 (18)\n",
    "@labeling_function()\n",
    "def lf_combined_modulator(x):\n",
    "    sentence_lower = x.text.lower()\n",
    "\n",
    "    specific_terms = ['allosteric modulator', 'positive modulator', 'negative modulator', 'non-competitive modulator', 'positive allosteric modulator']\n",
    "    if any(term in sentence_lower for term in specific_terms):\n",
    "        return 6\n",
    "\n",
    "    modulating_terms = ['modulat', 'allosteric', 'potentiate', 'regulate']\n",
    "    for term in modulating_terms:\n",
    "        if term in sentence_lower:\n",
    "            term_index = sentence_lower.index(term)\n",
    "            if x.entity1_index == -1 or x.entity2_index == -1:\n",
    "                return ABSTAIN\n",
    "            if abs(term_index - x.entity1_index) < 20 or abs(term_index - x.entity2_index) < 20:\n",
    "                return 6\n",
    "    # check the first indcidences of modulator and positive\n",
    "    if x.entity1 in sentence_lower and x.entity2 in sentence_lower:\n",
    "        entity1_index = sentence_lower.index(x.entity1.lower())\n",
    "        entity2_index = sentence_lower.index(x.entity2.lower())\n",
    "        between_entities = sentence_lower[min(entity1_index, entity2_index):max(entity1_index, entity2_index)]\n",
    "        if 'modulate' in between_entities or 'regulate' in between_entities:\n",
    "            return 6\n",
    "\n",
    "    if 'positive' in sentence_lower and 'modulator' in sentence_lower:\n",
    "        pos_indices = [i for i, word in enumerate(sentence_lower.split()) if word == 'positive']\n",
    "        mod_indices = [i for i, word in enumerate(sentence_lower.split()) if 'modulator' in word]\n",
    "        min_distance = min(abs(p - m) for p in pos_indices for m in mod_indices)\n",
    "        if min_distance <= 3:\n",
    "            return 6\n",
    "\n",
    "    return ABSTAIN\n",
    "\n",
    "lfs = [lf_amino_acid, lf_replace, lf_mutant, lf_bind, lf_interact, lf_affinity, lf_activate, lf_increase, lf_induce, lf_stimulate, lf_upregulate, lf_downregulate, lf_reduce, lf_inhibit, lf_decrease, lf_agonist, lf_antagonist, lf_combined_modulator, lf_allosteric, lf_cofactor, lf_substrate, lf_transport, lf_catalyze, lf_product, lf_convert, lf_not]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/1000 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1000/1000 [00:00<00:00, 7629.43it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "text             While SAR within the HTS series was very shall...\n",
      "labels                                                           6\n",
      "entity1                                                         M5\n",
      "entity2                                                      ML172\n",
      "span1                                                   [173, 175]\n",
      "span2                                                   [129, 134]\n",
      "weak_labels      [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...\n",
      "entity1_index                                                   28\n",
      "entity2_index                                                   21\n",
      "Name: 12446, dtype: object\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,\n",
       "       -1, -1, -1, -1, -1, -1, -1, -1, -1])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# show the first row of the dataframe\n",
    "train_dev = train_df.head(1000)\n",
    "train_dev = li.chemprot_enhanced(train_dev)\n",
    "L_dev = li.apply_LFs(lfs, train_dev)\n",
    "L_dev\n",
    "print(train_dev.iloc[352])\n",
    "L_dev[352]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test Coverage: 0.874\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>j</th>\n",
       "      <th>Polarity</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>Overlaps</th>\n",
       "      <th>Conflicts</th>\n",
       "      <th>Conflict Ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>lf_amino_acid</th>\n",
       "      <td>0</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_replace</th>\n",
       "      <td>1</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_mutant</th>\n",
       "      <td>2</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.018</td>\n",
       "      <td>0.620690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_bind</th>\n",
       "      <td>3</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.099</td>\n",
       "      <td>0.073</td>\n",
       "      <td>0.064</td>\n",
       "      <td>0.646465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_interact</th>\n",
       "      <td>4</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.578947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_affinity</th>\n",
       "      <td>5</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.043</td>\n",
       "      <td>0.032</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.604651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_activate</th>\n",
       "      <td>6</td>\n",
       "      <td>[2]</td>\n",
       "      <td>0.123</td>\n",
       "      <td>0.093</td>\n",
       "      <td>0.075</td>\n",
       "      <td>0.609756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_increase</th>\n",
       "      <td>7</td>\n",
       "      <td>[2]</td>\n",
       "      <td>0.111</td>\n",
       "      <td>0.074</td>\n",
       "      <td>0.063</td>\n",
       "      <td>0.567568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_induce</th>\n",
       "      <td>8</td>\n",
       "      <td>[2]</td>\n",
       "      <td>0.158</td>\n",
       "      <td>0.123</td>\n",
       "      <td>0.102</td>\n",
       "      <td>0.645570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_stimulate</th>\n",
       "      <td>9</td>\n",
       "      <td>[2]</td>\n",
       "      <td>0.043</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.030</td>\n",
       "      <td>0.697674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_upregulate</th>\n",
       "      <td>10</td>\n",
       "      <td>[2]</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.008</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_downregulate</th>\n",
       "      <td>11</td>\n",
       "      <td>[3]</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.018</td>\n",
       "      <td>0.017</td>\n",
       "      <td>0.772727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_reduce</th>\n",
       "      <td>12</td>\n",
       "      <td>[3]</td>\n",
       "      <td>0.079</td>\n",
       "      <td>0.065</td>\n",
       "      <td>0.053</td>\n",
       "      <td>0.670886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_inhibit</th>\n",
       "      <td>13</td>\n",
       "      <td>[3]</td>\n",
       "      <td>0.350</td>\n",
       "      <td>0.207</td>\n",
       "      <td>0.191</td>\n",
       "      <td>0.545714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_decrease</th>\n",
       "      <td>14</td>\n",
       "      <td>[3]</td>\n",
       "      <td>0.057</td>\n",
       "      <td>0.052</td>\n",
       "      <td>0.047</td>\n",
       "      <td>0.824561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_agonist</th>\n",
       "      <td>15</td>\n",
       "      <td>[4]</td>\n",
       "      <td>0.044</td>\n",
       "      <td>0.032</td>\n",
       "      <td>0.032</td>\n",
       "      <td>0.727273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_antagonist</th>\n",
       "      <td>16</td>\n",
       "      <td>[5]</td>\n",
       "      <td>0.089</td>\n",
       "      <td>0.061</td>\n",
       "      <td>0.061</td>\n",
       "      <td>0.685393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_combined_modulator</th>\n",
       "      <td>17</td>\n",
       "      <td>[6]</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.007</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_allosteric</th>\n",
       "      <td>18</td>\n",
       "      <td>[6]</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_cofactor</th>\n",
       "      <td>19</td>\n",
       "      <td>[7]</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_substrate</th>\n",
       "      <td>20</td>\n",
       "      <td>[8]</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.545455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_transport</th>\n",
       "      <td>21</td>\n",
       "      <td>[8]</td>\n",
       "      <td>0.036</td>\n",
       "      <td>0.023</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.583333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_catalyze</th>\n",
       "      <td>22</td>\n",
       "      <td>[8]</td>\n",
       "      <td>0.028</td>\n",
       "      <td>0.018</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.464286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_product</th>\n",
       "      <td>23</td>\n",
       "      <td>[8]</td>\n",
       "      <td>0.058</td>\n",
       "      <td>0.052</td>\n",
       "      <td>0.051</td>\n",
       "      <td>0.879310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_convert</th>\n",
       "      <td>24</td>\n",
       "      <td>[8]</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.533333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lf_not</th>\n",
       "      <td>25</td>\n",
       "      <td>[9]</td>\n",
       "      <td>0.093</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.881720</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        j Polarity  Coverage  Overlaps  Conflicts  \\\n",
       "lf_amino_acid           0      [0]     0.015     0.010      0.009   \n",
       "lf_replace              1      [0]     0.003     0.002      0.002   \n",
       "lf_mutant               2      [0]     0.029     0.019      0.018   \n",
       "lf_bind                 3      [1]     0.099     0.073      0.064   \n",
       "lf_interact             4      [1]     0.019     0.014      0.011   \n",
       "lf_affinity             5      [1]     0.043     0.032      0.026   \n",
       "lf_activate             6      [2]     0.123     0.093      0.075   \n",
       "lf_increase             7      [2]     0.111     0.074      0.063   \n",
       "lf_induce               8      [2]     0.158     0.123      0.102   \n",
       "lf_stimulate            9      [2]     0.043     0.036      0.030   \n",
       "lf_upregulate          10      [2]     0.008     0.008      0.008   \n",
       "lf_downregulate        11      [3]     0.022     0.018      0.017   \n",
       "lf_reduce              12      [3]     0.079     0.065      0.053   \n",
       "lf_inhibit             13      [3]     0.350     0.207      0.191   \n",
       "lf_decrease            14      [3]     0.057     0.052      0.047   \n",
       "lf_agonist             15      [4]     0.044     0.032      0.032   \n",
       "lf_antagonist          16      [5]     0.089     0.061      0.061   \n",
       "lf_combined_modulator  17      [6]     0.007     0.007      0.007   \n",
       "lf_allosteric          18      [6]     0.004     0.004      0.004   \n",
       "lf_cofactor            19      [7]     0.001     0.001      0.001   \n",
       "lf_substrate           20      [8]     0.022     0.014      0.012   \n",
       "lf_transport           21      [8]     0.036     0.023      0.021   \n",
       "lf_catalyze            22      [8]     0.028     0.018      0.013   \n",
       "lf_product             23      [8]     0.058     0.052      0.051   \n",
       "lf_convert             24      [8]     0.015     0.012      0.008   \n",
       "lf_not                 25      [9]     0.093     0.082      0.082   \n",
       "\n",
       "                       Conflict Ratio  \n",
       "lf_amino_acid                0.600000  \n",
       "lf_replace                   0.666667  \n",
       "lf_mutant                    0.620690  \n",
       "lf_bind                      0.646465  \n",
       "lf_interact                  0.578947  \n",
       "lf_affinity                  0.604651  \n",
       "lf_activate                  0.609756  \n",
       "lf_increase                  0.567568  \n",
       "lf_induce                    0.645570  \n",
       "lf_stimulate                 0.697674  \n",
       "lf_upregulate                1.000000  \n",
       "lf_downregulate              0.772727  \n",
       "lf_reduce                    0.670886  \n",
       "lf_inhibit                   0.545714  \n",
       "lf_decrease                  0.824561  \n",
       "lf_agonist                   0.727273  \n",
       "lf_antagonist                0.685393  \n",
       "lf_combined_modulator        1.000000  \n",
       "lf_allosteric                1.000000  \n",
       "lf_cofactor                  1.000000  \n",
       "lf_substrate                 0.545455  \n",
       "lf_transport                 0.583333  \n",
       "lf_catalyze                  0.464286  \n",
       "lf_product                   0.879310  \n",
       "lf_convert                   0.533333  \n",
       "lf_not                       0.881720  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Test Coverage:\", li.calc_coverage(L_dev))\n",
    "lf_analysis = LFAnalysis(L_dev, lfs).lf_summary()\n",
    "# Calculates how many of an LFs votes result in conflicts (helpful signal for debugging LFs)\n",
    "lf_analysis['Conflict Ratio'] = lf_analysis['Conflicts'] / lf_analysis['Coverage']\n",
    "lf_analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ 3  8 -1  8  3  1  8  2 -1  2  1 -1  3  3 -1  2  3  1  3 -1  3  1  1  3\n",
      "  1 -1  8 -1 -1  1 -1 -1  3  2  1 -1 -1 -1 -1  2  3  8 -1  5 -1  3  2  2\n",
      " -1  2 -1  3  3 -1  3  3  2 -1 -1  3  5 -1 -1  2  8 -1  3  2  3 -1 -1  3\n",
      "  2  3  2 -1 -1 -1  3  1  2  1  3 -1 -1 -1  3  3 -1 -1  2  8  3  3  3  1\n",
      " -1  8  2 -1  3 -1  1 -1 -1  3 -1 -1 -1 -1  0  3  3 -1 -1 -1  8 -1 -1 -1\n",
      "  3 -1 -1  1 -1  3 -1  3  2 -1  3  2 -1 -1  3  2  3 -1 -1  1  5  1  1 -1\n",
      " -1  3 -1  1  3  1  2  5  3 -1  5 -1 -1 -1 -1 -1 -1  2  9  3  8  4  3  3\n",
      " -1  3 -1 -1 -1 -1  9 -1 -1  2  2 -1  3  8  3  2 -1  8  3  2 -1  2  2 -1\n",
      " -1  8 -1  2 -1 -1  1 -1 -1 -1  3  2 -1  2  8 -1  3 -1 -1 -1 -1 -1  2 -1\n",
      "  2  5 -1  3  2 -1 -1  5  2 -1  2  1  2  2  2 -1  5  4  2 -1 -1 -1 -1 -1\n",
      "  1 -1 -1  3  2  0 -1  3 -1  3 -1 -1  3  8 -1  8  2 -1 -1  5 -1 -1 -1 -1\n",
      "  2 -1  3 -1  3  3 -1  3 -1  4 -1  8 -1 -1 -1 -1  3  1 -1 -1 -1  2 -1  3\n",
      " -1  1  3 -1 -1 -1  3 -1 -1  3 -1  8  8 -1  3  3 -1 -1 -1  1  1  2  3 -1\n",
      " -1  3  1  3  3 -1  3 -1  2  3  2 -1  8  5 -1  3 -1 -1  2  2  5  2 -1  2\n",
      " -1  3 -1  2  1 -1 -1 -1 -1  0  3 -1 -1 -1  0 -1 -1  9 -1 -1 -1 -1 -1  3\n",
      "  2 -1 -1 -1  3  4 -1 -1  8  2  3 -1 -1 -1 -1  8  3 -1 -1  0  3  3 -1  8\n",
      " -1  3  3 -1  0 -1 -1  2  3  3 -1  2 -1 -1  3 -1 -1  3  3 -1  3  4  2  2\n",
      "  2 -1 -1 -1 -1  3 -1  3 -1 -1  3  3  2 -1 -1 -1 -1  8 -1 -1  2  2 -1 -1\n",
      "  8 -1 -1  4 -1  5  1 -1 -1  1  3 -1 -1 -1  3 -1  2  9  2  3  2  2  8  3\n",
      "  2 -1  3  0 -1  2 -1  2  8 -1  3 -1 -1  5  3  2  2 -1  3  3 -1 -1 -1 -1\n",
      " -1 -1  2 -1 -1  3  3 -1  8  3 -1 -1 -1  3  1 -1 -1  2 -1  8 -1 -1  3  3\n",
      "  2 -1  8  1 -1  2  8 -1  3  0 -1  3 -1  5  8  2  2  3  3  3  0 -1  8 -1\n",
      " -1 -1 -1  2  1 -1  2  3  2 -1  3  8  3 -1  0 -1 -1  5  3  1  8  2 -1  2\n",
      "  1 -1 -1  1 -1 -1 -1  9  2 -1 -1  2 -1  3 -1 -1  1 -1 -1 -1 -1  5 -1  3\n",
      "  2 -1 -1 -1  3 -1  1 -1  3  3 -1  8 -1  3  1  3 -1  3  3 -1  3  2  2 -1\n",
      " -1 -1  2  3 -1 -1  8 -1  3  2  3  3  1 -1  2 -1 -1  0 -1  2 -1 -1  1  5\n",
      "  3 -1  3  3  2  5  2  3 -1  1  2 -1 -1  3  9  5  2  3 -1 -1 -1  8  3 -1\n",
      "  2  2 -1  2  3 -1  0  8 -1  8 -1  3 -1 -1  1  3  2  2 -1 -1 -1  3  6  3\n",
      " -1 -1 -1  2 -1  2 -1  2  3 -1  3  5  3  2  3 -1  2  1  8  3  2  9  2  2\n",
      "  0 -1  5 -1  2  8  3  3 -1  2 -1 -1  2  3  3 -1 -1  3 -1  1  2  2  5  3\n",
      "  4 -1 -1 -1 -1  8  3 -1 -1 -1 -1 -1 -1  2  2  2  3  2  1 -1  2 -1 -1  8\n",
      "  3 -1 -1  2  2  2 -1  8 -1  3  2  3  2  3  3  2  1  2 -1  9 -1  2  9  8\n",
      " -1 -1  2 -1  2  0 -1 -1  3  2 -1  9 -1  1 -1  4 -1  2 -1  5 -1  3 -1  2\n",
      "  4  1  1 -1  2  0  2  3  2  3  3  5  3  2  1  3  3  2 -1 -1  3  3  2  3\n",
      "  2 -1 -1  2 -1  5 -1  2 -1 -1 -1  8  4  6  3 -1  3 -1 -1  2  3 -1 -1 -1\n",
      " -1  2  3 -1 -1  2  3  3  2  3  3  8 -1 -1  8  3 -1  4  2 -1 -1 -1  1  3\n",
      "  3 -1 -1  2  1  8  3  2  3  3  2  1 -1  2  2 -1  3  2  1  2 -1 -1  8  2\n",
      " -1 -1  2  3  8 -1  2  2 -1  3  2  2 -1  8  2  3  8 -1  2  2  1 -1  2 -1\n",
      "  3 -1 -1 -1 -1  0  2  3  2 -1  1 -1 -1  1 -1  3 -1  2  8 -1  2 -1  2  5\n",
      "  3  3 -1 -1  2 -1  2  3  2 -1 -1 -1 -1  3  2  2  0  9  3 -1 -1  3  2  1\n",
      " -1 -1 -1  2 -1  3 -1  2 -1  1 -1 -1 -1  3 -1 -1  3  3  8  0  1  3 -1 -1\n",
      "  3  3  3 -1  5  3  2  3  4  2  2  5  3 -1 -1  3]\n",
      "-1\n",
      "0.6602787456445993\n",
      "text             While SAR within the HTS series was very shall...\n",
      "labels                                                           6\n",
      "entity1                                                         M5\n",
      "entity2                                                      ML172\n",
      "span1                                                   [173, 175]\n",
      "span2                                                   [129, 134]\n",
      "weak_labels      [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...\n",
      "entity1_index                                                   28\n",
      "entity2_index                                                   21\n",
      "Name: 12446, dtype: object\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>text</th>\n",
       "      <th>labels</th>\n",
       "      <th>entity1</th>\n",
       "      <th>entity2</th>\n",
       "      <th>span1</th>\n",
       "      <th>span2</th>\n",
       "      <th>weak_labels</th>\n",
       "      <th>entity1_index</th>\n",
       "      <th>entity2_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12446</th>\n",
       "      <td>While SAR within the HTS series was very shall...</td>\n",
       "      <td>6</td>\n",
       "      <td>M5</td>\n",
       "      <td>ML172</td>\n",
       "      <td>[173, 175]</td>\n",
       "      <td>[129, 134]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...</td>\n",
       "      <td>28.0</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    text  labels entity1  \\\n",
       "12446  While SAR within the HTS series was very shall...       6      M5   \n",
       "\n",
       "      entity2       span1       span2  \\\n",
       "12446   ML172  [173, 175]  [129, 134]   \n",
       "\n",
       "                                             weak_labels  entity1_index  \\\n",
       "12446  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...           28.0   \n",
       "\n",
       "       entity2_index  \n",
       "12446           21.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate accuracy on the validation set (Ideally do this only at the end)\n",
    "majority_model = MajorityLabelVoter(10)\n",
    "preds_valid = majority_model.predict(L=L_dev)\n",
    "print(preds_valid)\n",
    "print(preds_valid[352])\n",
    "print((preds_valid[preds_valid != -1] == train_dev[preds_valid != -1].labels.values).mean())\n",
    "print(train_dev.iloc[352])\n",
    "incorrect_indices = np.where(preds_valid[preds_valid != -1] != train_dev[preds_valid != -1].labels.values)[0]\n",
    "incorrect_predictions_df = train_dev.iloc[incorrect_indices]\n",
    "label_six_df = incorrect_predictions_df[incorrect_predictions_df.labels == 6]\n",
    "label_six_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/1000 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1000/1000 [00:00<00:00, 8209.08it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[-1, -1, -1, ..., -1, -1, -1],\n",
       "       [-1, -1, -1, ..., -1, -1, -1],\n",
       "       [-1, -1, -1, ..., -1, -1, -1],\n",
       "       ...,\n",
       "       [-1, -1, -1, ..., -1, -1, -1],\n",
       "       [-1, -1, -1, ..., -1, -1, -1],\n",
       "       [-1, -1, -1, ..., -1, -1, -1]])"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lfs = [lf_amino_acid, lf_replace, lf_mutant, lf_bind, lf_interact, lf_affinity, lf_activate, lf_increase, lf_induce, lf_stimulate, lf_upregulate, lf_downregulate, lf_reduce, lf_inhibit, lf_decrease, lf_agonist, lf_antagonist, lf_modulate, lf_allosteric, lf_cofactor, lf_substrate, lf_transport, lf_catalyze, lf_product, lf_convert, lf_not]\n",
    "\n",
    "# show the first row of the dataframe\n",
    "train_dev = train_df.head(1000)\n",
    "train_dev = li.chemprot_enhanced(train_dev)\n",
    "L_dev2 = li.apply_LFs(lfs, train_dev)\n",
    "L_dev2  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "difference = np.where(L_dev2 !=  L_dev)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "text             This effect appeared to be due to both competi...\n",
      "labels                                                           1\n",
      "entity1                                                        Trx\n",
      "entity2                                          S-nitrosocysteine\n",
      "span1                                                     [94, 97]\n",
      "span2                                                     [59, 76]\n",
      "weak_labels      [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...\n",
      "entity1_index                                                   15\n",
      "entity2_index                                                   10\n",
      "Name: 15523, dtype: object\n",
      "[-1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1  3 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1\n",
      " -1 -1]\n",
      "[-1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1  3 -1 -1 -1 -1  6 -1 -1 -1 -1 -1 -1\n",
      " -1 -1]\n",
      "\n",
      "\n",
      "text             Thalidomide--removed from widespread clinical ...\n",
      "labels                                                           3\n",
      "entity1                                                  TNF alpha\n",
      "entity2                                                Thalidomide\n",
      "span1                                                   [170, 179]\n",
      "span2                                                      [0, 11]\n",
      "weak_labels      [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -...\n",
      "entity1_index                                                 19.5\n",
      "entity2_index                                                    0\n",
      "Name: 6532, dtype: object\n",
      "[-1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1  3 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1\n",
      " -1 -1]\n",
      "[-1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1  3 -1 -1 -1  6 -1 -1 -1 -1 -1 -1\n",
      " -1 -1]\n",
      "\n",
      "\n",
      "text             DBDCT up-regulated the expression of Bax, down...\n",
      "labels                                                           2\n",
      "entity1                                                        Bax\n",
      "entity2                                                      DBDCT\n",
      "span1                                                   [123, 126]\n",
      "span2                                                       [0, 5]\n",
      "weak_labels      [-1, -1, -1, -1, -1, -1, -1, 2, -1, -1, -1, 3,...\n",
      "entity1_index                                                   17\n",
      "entity2_index                                                    0\n",
      "Name: 3649, dtype: object\n",
      "[-1 -1 -1 -1 -1 -1 -1  2 -1 -1 -1  3 -1 -1 -1 -1 -1  6 -1 -1 -1 -1 -1 -1\n",
      " -1 -1]\n",
      "[-1 -1 -1 -1 -1 -1 -1  2 -1 -1 -1  3 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1\n",
      " -1 -1]\n",
      "\n",
      "\n",
      "text             This study is the first to identify Class V CG...\n",
      "labels                                                           1\n",
      "entity1                                                        MRP\n",
      "entity2                                                    methine\n",
      "span1                                                   [180, 183]\n",
      "span2                                                     [72, 79]\n",
      "weak_labels      [-1, -1, -1, -1, -1, -1, -1, -1, -1, 2, -1, -1...\n",
      "entity1_index                                                   28\n",
      "entity2_index                                                   13\n",
      "Name: 5005, dtype: object\n",
      "[-1 -1 -1 -1 -1 -1 -1 -1 -1  2 -1 -1 -1  3 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1\n",
      " -1 -1]\n",
      "[-1 -1 -1 -1 -1 -1 -1 -1 -1  2 -1 -1 -1  3 -1 -1 -1  6 -1 -1 -1 -1 -1 -1\n",
      " -1 -1]\n",
      "\n",
      "\n",
      "text             We previously demonstrated that the somatostat...\n",
      "labels                                                           2\n",
      "entity1          cyclic adenosine monophosphate response elemen...\n",
      "entity2                                               α-tocopherol\n",
      "span1                                                   [173, 236]\n",
      "span2                                                   [123, 135]\n",
      "weak_labels      [-1, -1, -1, 1, -1, -1, -1, 2, -1, -1, -1, -1,...\n",
      "entity1_index                                                   26\n",
      "entity2_index                                                   16\n",
      "Name: 14189, dtype: object\n",
      "[-1 -1 -1  1 -1 -1 -1  2 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1\n",
      " -1 -1]\n",
      "[-1 -1 -1  1 -1 -1 -1  2 -1 -1 -1 -1 -1 -1 -1 -1 -1  6 -1 -1 -1 -1 -1 -1\n",
      " -1 -1]\n",
      "\n",
      "\n",
      "text             Diacylglycerol (DAG) acts as an allosteric act...\n",
      "labels                                                           8\n",
      "entity1                                                        DGK\n",
      "entity2                                          phosphatidic acid\n",
      "span1                                                   [132, 135]\n",
      "span2                                                    [99, 116]\n",
      "weak_labels      [-1, -1, -1, -1, -1, -1, 2, -1, -1, -1, -1, -1...\n",
      "entity1_index                                                   21\n",
      "entity2_index                                                 16.5\n",
      "Name: 12980, dtype: object\n",
      "[-1 -1 -1 -1 -1 -1  2 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1  6  6 -1 -1 -1 -1 -1\n",
      "  8 -1]\n",
      "[-1 -1 -1 -1 -1 -1  2 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1  6 -1 -1 -1 -1 -1\n",
      "  8 -1]\n",
      "\n",
      "\n",
      "text             Lisuride produced a significant dose-related i...\n",
      "labels                                                           1\n",
      "entity1                                                   5-HT(1A)\n",
      "entity2                                                   Lisuride\n",
      "span1                                                   [179, 187]\n",
      "span2                                                       [0, 8]\n",
      "weak_labels      [-1, -1, -1, -1, -1, -1, -1, 2, -1, -1, -1, -1...\n",
      "entity1_index                                                   25\n",
      "entity2_index                                                    0\n",
      "Name: 1414, dtype: object\n",
      "[-1 -1 -1 -1 -1 -1 -1  2 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1  8\n",
      " -1 -1]\n",
      "[-1 -1 -1 -1 -1 -1 -1  2 -1 -1 -1 -1 -1 -1 -1 -1 -1  6 -1 -1 -1 -1 -1  8\n",
      " -1 -1]\n",
      "\n",
      "\n",
      "text             Similarly, pranlukast did not modulate IL-5- o...\n",
      "labels                                                           9\n",
      "entity1                                                       IL-5\n",
      "entity2                                                 pranlukast\n",
      "span1                                                     [39, 43]\n",
      "span2                                                     [11, 21]\n",
      "weak_labels      [-1, -1, -1, -1, -1, -1, 2, -1, -1, -1, -1, -1...\n",
      "entity1_index                                                    5\n",
      "entity2_index                                                    1\n",
      "Name: 6629, dtype: object\n",
      "[-1 -1 -1 -1 -1 -1  2 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1\n",
      " -1  9]\n",
      "[-1 -1 -1 -1 -1 -1  2 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1  6 -1 -1 -1 -1 -1 -1\n",
      " -1  9]\n",
      "\n",
      "\n",
      "text             Additionally, MPTP significantly down-regulate...\n",
      "labels                                                           2\n",
      "entity1                                                        Bax\n",
      "entity2                                                       MPTP\n",
      "span1                                                   [145, 148]\n",
      "span2                                                     [14, 18]\n",
      "weak_labels      [-1, -1, -1, -1, -1, -1, -1, 2, -1, -1, -1, 3,...\n",
      "entity1_index                                                   20\n",
      "entity2_index                                                    1\n",
      "Name: 5424, dtype: object\n",
      "[-1 -1 -1 -1 -1 -1 -1  2 -1 -1 -1  3 -1  3 -1 -1 -1  6 -1 -1 -1 -1 -1 -1\n",
      " -1 -1]\n",
      "[-1 -1 -1 -1 -1 -1 -1  2 -1 -1 -1  3 -1  3 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1\n",
      " -1 -1]\n",
      "\n",
      "\n",
      "9\n"
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "for i in difference:\n",
    "    print(train_dev.iloc[i])\n",
    "    print(train_dev.iloc[i].text)\n",
    "    print(L_dev[i])\n",
    "    print(L_dev2[i])\n",
    "    count +=1\n",
    "    print(\"\\n\")\n",
    "print(count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate accuracy on the validation set (Ideally do this only at the end)\n",
    "majority_model = MajorityLabelVoter(10)\n",
    "preds_valid = majority_model.predict(L=L_dev)\n",
    "print(preds_valid)\n",
    "print(preds_valid[352])\n",
    "print((preds_valid[preds_valid != -1] == train_dev[preds_valid != -1].labels.values).mean())\n",
    "print(train_dev.iloc[352])\n",
    "incorrect_indices = np.where(preds_valid[preds_valid != -1] != train_dev[preds_valid != -1].labels.values)[0]\n",
    "incorrect_predictions_df = train_dev.iloc[incorrect_indices]\n",
    "label_six_df = incorrect_predictions_df[incorrect_predictions_df.labels == 6]\n",
    "label_six_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DO NOT GO BEYOND THIS LINE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ------------------------------------------------------------------------------------ #"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>text</th>\n",
       "      <th>labels</th>\n",
       "      <th>entity1</th>\n",
       "      <th>entity2</th>\n",
       "      <th>span1</th>\n",
       "      <th>span2</th>\n",
       "      <th>weak_labels</th>\n",
       "      <th>entity1_index</th>\n",
       "      <th>entity2_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10305</th>\n",
       "      <td>Selective inhibition of PDE5 is a rational therapeutic approach in ED, as proved by the clinical success of sildenafil.</td>\n",
       "      <td>3</td>\n",
       "      <td>PDE5</td>\n",
       "      <td>sildenafil</td>\n",
       "      <td>[24, 28]</td>\n",
       "      <td>[108, 118]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>3.0</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8422</th>\n",
       "      <td>Furthermore, knockdown of OPN enhanced cell death caused by other drugs, including paclitaxel, doxorubicin, actinomycin-D, and rapamycin, which are also P-gp substrates.</td>\n",
       "      <td>8</td>\n",
       "      <td>P-gp</td>\n",
       "      <td>paclitaxel</td>\n",
       "      <td>[153, 157]</td>\n",
       "      <td>[83, 93]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 8, -1, -1, -1, -1, -1]</td>\n",
       "      <td>20.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8486</th>\n",
       "      <td>Furthermore, no impact on cytokine release (i.e., on IL-10, IL-6, IL-12/23p40 and TNFα levels) was seen in LPS-stimulated human PBMCs, except with JWH-210 and JWH-122 which caused a decrease of TNFα and IL-12/23p40.</td>\n",
       "      <td>3</td>\n",
       "      <td>IL-12</td>\n",
       "      <td>JWH-210</td>\n",
       "      <td>[203, 208]</td>\n",
       "      <td>[147, 154]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, 2, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>32.0</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6254</th>\n",
       "      <td>Among neuroleptics, the four most potent compounds at the human serotonin transporter were triflupromazine, fluperlapine, chlorpromazine, and ziprasidone (K(D) 24-39 nM); and at the norepinephrine transporter, chlorpromazine, zotepine, chlorprothixene, and promazine (K(D) 19-25 nM).</td>\n",
       "      <td>1</td>\n",
       "      <td>norepinephrine transporter</td>\n",
       "      <td>chlorpromazine</td>\n",
       "      <td>[182, 208]</td>\n",
       "      <td>[210, 224]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 8, -1, -1, -1, -1]</td>\n",
       "      <td>24.5</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1801</th>\n",
       "      <td>Epidermal growth factor receptor inhibitors currently under investigation include the small molecules gefitinib (Iressa, ZD1839) and erlotinib (Tarceva, OSI-774), as well as monoclonal antibodies such as cetuximab (IMC-225, Erbitux).</td>\n",
       "      <td>3</td>\n",
       "      <td>Epidermal growth factor receptor</td>\n",
       "      <td>Tarceva</td>\n",
       "      <td>[0, 32]</td>\n",
       "      <td>[144, 151]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>1.5</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9350</th>\n",
       "      <td>Tamsulosin, the first prostate-selective alpha 1A-adrenoceptor antagonist.</td>\n",
       "      <td>5</td>\n",
       "      <td>alpha 1A-adrenoceptor</td>\n",
       "      <td>Tamsulosin</td>\n",
       "      <td>[41, 62]</td>\n",
       "      <td>[0, 10]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 5, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>4.5</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13234</th>\n",
       "      <td>The GRIP1 reduction was inhibited by MK-801, an N-methyl-d-aspartate (NMDA) receptor antagonist, but not by 6-cyano-7-nitroquinoxaline-2,3-dione (CNQX), an AMPA receptor antagonist.</td>\n",
       "      <td>2</td>\n",
       "      <td>GRIP1</td>\n",
       "      <td>MK-801</td>\n",
       "      <td>[4, 9]</td>\n",
       "      <td>[37, 43]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, 3, -1, -1, 5, -1, -1, -1, -1, -1, -1, -1, -1, 9]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1837</th>\n",
       "      <td>The reciprocal inhibition of SR-BI and ABCA1 by BLT-4 and glyburide raises the possibility that these proteins may share similar or common steps in their mechanisms of lipid transport.</td>\n",
       "      <td>3</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>glyburide</td>\n",
       "      <td>[39, 44]</td>\n",
       "      <td>[58, 67]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, 8, -1, -1, -1, -1]</td>\n",
       "      <td>6.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11290</th>\n",
       "      <td>Firstly, transgenic plants overexpressing formate dehydrogenase (FDH, EC 1.2.1.2) were used to continue our previous studies on the function of FDH in formate metabolism.</td>\n",
       "      <td>8</td>\n",
       "      <td>EC 1.2.1.2</td>\n",
       "      <td>formate</td>\n",
       "      <td>[70, 80]</td>\n",
       "      <td>[151, 158]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>7.5</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5424</th>\n",
       "      <td>Additionally, MPTP significantly down-regulated Bcl-2 expression in the mitochondria of dopaminergic cells in the SN, followed by an increase in Bax expression, cytochrome C translocation to the cytosol, andcleaved-caspase-3 expression, whereas these were inhibited by CRE or EB treatment.</td>\n",
       "      <td>2</td>\n",
       "      <td>Bax</td>\n",
       "      <td>MPTP</td>\n",
       "      <td>[145, 148]</td>\n",
       "      <td>[14, 18]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, 2, -1, -1, -1, 3, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                                                                                                                                    text  \\\n",
       "10305                                                                                                                                                                            Selective inhibition of PDE5 is a rational therapeutic approach in ED, as proved by the clinical success of sildenafil.   \n",
       "8422                                                                                                                           Furthermore, knockdown of OPN enhanced cell death caused by other drugs, including paclitaxel, doxorubicin, actinomycin-D, and rapamycin, which are also P-gp substrates.   \n",
       "8486                                                                             Furthermore, no impact on cytokine release (i.e., on IL-10, IL-6, IL-12/23p40 and TNFα levels) was seen in LPS-stimulated human PBMCs, except with JWH-210 and JWH-122 which caused a decrease of TNFα and IL-12/23p40.   \n",
       "6254         Among neuroleptics, the four most potent compounds at the human serotonin transporter were triflupromazine, fluperlapine, chlorpromazine, and ziprasidone (K(D) 24-39 nM); and at the norepinephrine transporter, chlorpromazine, zotepine, chlorprothixene, and promazine (K(D) 19-25 nM).   \n",
       "1801                                                           Epidermal growth factor receptor inhibitors currently under investigation include the small molecules gefitinib (Iressa, ZD1839) and erlotinib (Tarceva, OSI-774), as well as monoclonal antibodies such as cetuximab (IMC-225, Erbitux).   \n",
       "...                                                                                                                                                                                                                                                                                                  ...   \n",
       "9350                                                                                                                                                                                                                          Tamsulosin, the first prostate-selective alpha 1A-adrenoceptor antagonist.   \n",
       "13234                                                                                                              The GRIP1 reduction was inhibited by MK-801, an N-methyl-d-aspartate (NMDA) receptor antagonist, but not by 6-cyano-7-nitroquinoxaline-2,3-dione (CNQX), an AMPA receptor antagonist.   \n",
       "1837                                                                                                            The reciprocal inhibition of SR-BI and ABCA1 by BLT-4 and glyburide raises the possibility that these proteins may share similar or common steps in their mechanisms of lipid transport.   \n",
       "11290                                                                                                                         Firstly, transgenic plants overexpressing formate dehydrogenase (FDH, EC 1.2.1.2) were used to continue our previous studies on the function of FDH in formate metabolism.   \n",
       "5424   Additionally, MPTP significantly down-regulated Bcl-2 expression in the mitochondria of dopaminergic cells in the SN, followed by an increase in Bax expression, cytochrome C translocation to the cytosol, andcleaved-caspase-3 expression, whereas these were inhibited by CRE or EB treatment.   \n",
       "\n",
       "       labels                           entity1         entity2       span1  \\\n",
       "10305       3                              PDE5      sildenafil    [24, 28]   \n",
       "8422        8                              P-gp      paclitaxel  [153, 157]   \n",
       "8486        3                             IL-12         JWH-210  [203, 208]   \n",
       "6254        1        norepinephrine transporter  chlorpromazine  [182, 208]   \n",
       "1801        3  Epidermal growth factor receptor         Tarceva     [0, 32]   \n",
       "...       ...                               ...             ...         ...   \n",
       "9350        5             alpha 1A-adrenoceptor      Tamsulosin    [41, 62]   \n",
       "13234       2                             GRIP1          MK-801      [4, 9]   \n",
       "1837        3                             ABCA1       glyburide    [39, 44]   \n",
       "11290       8                        EC 1.2.1.2         formate    [70, 80]   \n",
       "5424        2                               Bax            MPTP  [145, 148]   \n",
       "\n",
       "            span2  \\\n",
       "10305  [108, 118]   \n",
       "8422     [83, 93]   \n",
       "8486   [147, 154]   \n",
       "6254   [210, 224]   \n",
       "1801   [144, 151]   \n",
       "...           ...   \n",
       "9350      [0, 10]   \n",
       "13234    [37, 43]   \n",
       "1837     [58, 67]   \n",
       "11290  [151, 158]   \n",
       "5424     [14, 18]   \n",
       "\n",
       "                                                                                                    weak_labels  \\\n",
       "10305   [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "8422    [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 8, -1, -1, -1, -1, -1]   \n",
       "8486     [-1, -1, -1, -1, -1, -1, -1, -1, -1, 2, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "6254    [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 8, -1, -1, -1, -1]   \n",
       "1801    [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "...                                                                                                         ...   \n",
       "9350    [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 5, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "13234      [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, 3, -1, -1, 5, -1, -1, -1, -1, -1, -1, -1, -1, 9]   \n",
       "1837     [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, 8, -1, -1, -1, -1]   \n",
       "11290  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "5424      [-1, -1, -1, -1, -1, -1, -1, 2, -1, -1, -1, 3, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "\n",
       "       entity1_index  entity2_index  \n",
       "10305            3.0           18.0  \n",
       "8422            20.0           12.0  \n",
       "8486            32.0           22.0  \n",
       "6254            24.5           26.0  \n",
       "1801             1.5           17.0  \n",
       "...              ...            ...  \n",
       "9350             4.5            0.0  \n",
       "13234            1.0            6.0  \n",
       "1837             6.0           10.0  \n",
       "11290            7.5           22.0  \n",
       "5424            20.0            1.0  \n",
       "\n",
       "[1000 rows x 9 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = li.chemprot_enhanced(train_dev)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1000/1000 [00:00<00:00, 66886.27it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [ 6],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [ 6],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [ 6],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [ 6],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [ 6],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [ 6],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [ 6]])"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lfs = [lf_combined_modulator]\n",
    "\n",
    "# Initialize the applier\n",
    "applier = PandasLFApplier(lfs=lfs)\n",
    "L_train = applier.apply(df=df)\n",
    "L_train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>text</th>\n",
       "      <th>labels</th>\n",
       "      <th>entity1</th>\n",
       "      <th>entity2</th>\n",
       "      <th>span1</th>\n",
       "      <th>span2</th>\n",
       "      <th>weak_labels</th>\n",
       "      <th>entity1_index</th>\n",
       "      <th>entity2_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12446</th>\n",
       "      <td>While SAR within the HTS series was very shallow and unable to be optimized, grafting the phenethyl ether linkage onto the ML129/ML172 cores led to the first sub-micromolar M5 PAM, ML326 (VU0467903), (human and rat M5 EC50s of 409nM and 500nM, respectively) with excellent mAChR selectivity (M1-M4 EC50s &gt;30μM) and a robust 20-fold leftward shift of the ACh CRC.</td>\n",
       "      <td>6</td>\n",
       "      <td>M5</td>\n",
       "      <td>ML172</td>\n",
       "      <td>[173, 175]</td>\n",
       "      <td>[129, 134]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>28.0</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3477</th>\n",
       "      <td>Anxiolytic- but not antidepressant-like activity of Lu AF21934, a novel, selective positive allosteric modulator of the mGlu₄ receptor.</td>\n",
       "      <td>6</td>\n",
       "      <td>mGlu₄</td>\n",
       "      <td>Lu AF21934</td>\n",
       "      <td>[120, 125]</td>\n",
       "      <td>[52, 62]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, 6, -1, -1, -1, -1, -1, -1, 9]</td>\n",
       "      <td>16.0</td>\n",
       "      <td>6.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8655</th>\n",
       "      <td>Conversely, ovarian PRA and PRB were positively regulated by ethanol and ethanol-melatonin combination, whereas PRA was down-regulated in the uterus and oviduct after ethanol consumption.</td>\n",
       "      <td>6</td>\n",
       "      <td>PRB</td>\n",
       "      <td>melatonin</td>\n",
       "      <td>[28, 31]</td>\n",
       "      <td>[81, 90]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 8, -1]</td>\n",
       "      <td>4.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3387</th>\n",
       "      <td>BDZs and other positive GABA(A)R modulators, including barbiturates, ethanol, and neurosteroids, can also inhibit L-type voltage-gated calcium channels (L-VGCCs), which could contribute to reduced neuronal excitability.</td>\n",
       "      <td>6</td>\n",
       "      <td>GABA(A)R</td>\n",
       "      <td>BDZs</td>\n",
       "      <td>[24, 32]</td>\n",
       "      <td>[0, 4]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, 3, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8654</th>\n",
       "      <td>Conversely, ovarian PRA and PRB were positively regulated by ethanol and ethanol-melatonin combination, whereas PRA was down-regulated in the uterus and oviduct after ethanol consumption.</td>\n",
       "      <td>6</td>\n",
       "      <td>PRA</td>\n",
       "      <td>melatonin</td>\n",
       "      <td>[20, 23]</td>\n",
       "      <td>[81, 90]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 8, -1]</td>\n",
       "      <td>2.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13836</th>\n",
       "      <td>The recently discovered hyperinsulinism/hyperammonemia disorder showed that the loss of allosteric inhibition of GDH by GTP causes excessive secretion of insulin.</td>\n",
       "      <td>6</td>\n",
       "      <td>GDH</td>\n",
       "      <td>GTP</td>\n",
       "      <td>[113, 116]</td>\n",
       "      <td>[120, 123]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>13.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5378</th>\n",
       "      <td>The implications of our finding for mechanism of in vivo actions of rapamycin and for design of novel allosteric drugs targeting the proteasome are discussed.</td>\n",
       "      <td>6</td>\n",
       "      <td>proteasome</td>\n",
       "      <td>rapamycin</td>\n",
       "      <td>[133, 143]</td>\n",
       "      <td>[68, 77]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>22.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3606</th>\n",
       "      <td>Ifenprodil is an allosteric inhibitor of GluN1/GluN2B N-methyl-D-aspartate receptors.</td>\n",
       "      <td>6</td>\n",
       "      <td>GluN1</td>\n",
       "      <td>Ifenprodil</td>\n",
       "      <td>[41, 46]</td>\n",
       "      <td>[0, 10]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15281</th>\n",
       "      <td>Synthesis and structure-activity relationships of indazole arylsulfonamides as allosteric CC-chemokine receptor 4 (CCR4) antagonists.</td>\n",
       "      <td>6</td>\n",
       "      <td>CC-chemokine receptor 4</td>\n",
       "      <td>indazole arylsulfonamides</td>\n",
       "      <td>[90, 113]</td>\n",
       "      <td>[50, 75]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 5, -1, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>10.0</td>\n",
       "      <td>5.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3381</th>\n",
       "      <td>Benzodiazepines (BDZs) depress neuronal excitability via positive allosteric modulation of inhibitory GABA(A) receptors (GABA(A)R).</td>\n",
       "      <td>6</td>\n",
       "      <td>GABA(A)R</td>\n",
       "      <td>BDZs</td>\n",
       "      <td>[121, 129]</td>\n",
       "      <td>[17, 21]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, 6, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14216</th>\n",
       "      <td>The structurally diverse opioids codeine and eseroline, like galantamine, are also nAChR-APL that have greatly diminished affinity for AChE, representing potential lead compounds for selective nAChR-APL development.</td>\n",
       "      <td>6</td>\n",
       "      <td>nAChR</td>\n",
       "      <td>eseroline</td>\n",
       "      <td>[193, 198]</td>\n",
       "      <td>[45, 54]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>25.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8650</th>\n",
       "      <td>Conversely, ovarian PRA and PRB were positively regulated by ethanol and ethanol-melatonin combination, whereas PRA was down-regulated in the uterus and oviduct after ethanol consumption.</td>\n",
       "      <td>6</td>\n",
       "      <td>PRA</td>\n",
       "      <td>ethanol</td>\n",
       "      <td>[20, 23]</td>\n",
       "      <td>[61, 68]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 8, -1]</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12445</th>\n",
       "      <td>While SAR within the HTS series was very shallow and unable to be optimized, grafting the phenethyl ether linkage onto the ML129/ML172 cores led to the first sub-micromolar M5 PAM, ML326 (VU0467903), (human and rat M5 EC50s of 409nM and 500nM, respectively) with excellent mAChR selectivity (M1-M4 EC50s &gt;30μM) and a robust 20-fold leftward shift of the ACh CRC.</td>\n",
       "      <td>6</td>\n",
       "      <td>M5</td>\n",
       "      <td>ML129</td>\n",
       "      <td>[173, 175]</td>\n",
       "      <td>[123, 128]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>28.0</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9857</th>\n",
       "      <td>Aspirin and sodium salicylate inhibit endothelin ETA receptors by an allosteric type of mechanism.</td>\n",
       "      <td>6</td>\n",
       "      <td>ETA receptors</td>\n",
       "      <td>Aspirin</td>\n",
       "      <td>[49, 62]</td>\n",
       "      <td>[0, 7]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>6.5</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13869</th>\n",
       "      <td>Because these speeds are significantly faster than the responses to antagonists, these data indicate that gallamine and dimethyl-W84 are allosteric ligands and actively induce a conformation of the M(2) receptor with a reduced affinity for its agonists.</td>\n",
       "      <td>6</td>\n",
       "      <td>M(2) receptor</td>\n",
       "      <td>gallamine</td>\n",
       "      <td>[198, 211]</td>\n",
       "      <td>[106, 115]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, 1, -1, -1, 2, -1, -1, -1, 3, -1, -1, 4, 5, -1, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>28.5</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8651</th>\n",
       "      <td>Conversely, ovarian PRA and PRB were positively regulated by ethanol and ethanol-melatonin combination, whereas PRA was down-regulated in the uterus and oviduct after ethanol consumption.</td>\n",
       "      <td>6</td>\n",
       "      <td>PRB</td>\n",
       "      <td>ethanol</td>\n",
       "      <td>[28, 31]</td>\n",
       "      <td>[61, 68]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 8, -1]</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5381</th>\n",
       "      <td>Rapamycin is a canonical allosteric inhibitor of the mTOR kinase with immunosuppressive and pro-apoptotic activities.</td>\n",
       "      <td>6</td>\n",
       "      <td>kinase</td>\n",
       "      <td>Rapamycin</td>\n",
       "      <td>[58, 64]</td>\n",
       "      <td>[0, 9]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12448</th>\n",
       "      <td>While SAR within the HTS series was very shallow and unable to be optimized, grafting the phenethyl ether linkage onto the ML129/ML172 cores led to the first sub-micromolar M5 PAM, ML326 (VU0467903), (human and rat M5 EC50s of 409nM and 500nM, respectively) with excellent mAChR selectivity (M1-M4 EC50s &gt;30μM) and a robust 20-fold leftward shift of the ACh CRC.</td>\n",
       "      <td>6</td>\n",
       "      <td>M5</td>\n",
       "      <td>ML326</td>\n",
       "      <td>[215, 217]</td>\n",
       "      <td>[181, 186]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>35.0</td>\n",
       "      <td>30.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12447</th>\n",
       "      <td>While SAR within the HTS series was very shallow and unable to be optimized, grafting the phenethyl ether linkage onto the ML129/ML172 cores led to the first sub-micromolar M5 PAM, ML326 (VU0467903), (human and rat M5 EC50s of 409nM and 500nM, respectively) with excellent mAChR selectivity (M1-M4 EC50s &gt;30μM) and a robust 20-fold leftward shift of the ACh CRC.</td>\n",
       "      <td>6</td>\n",
       "      <td>M5</td>\n",
       "      <td>ML326</td>\n",
       "      <td>[173, 175]</td>\n",
       "      <td>[181, 186]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>28.0</td>\n",
       "      <td>30.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12978</th>\n",
       "      <td>Diacylglycerol (DAG) acts as an allosteric activator of protein kinase C (PKC) and is converted to phosphatidic acid by DAG kinase (DGK).</td>\n",
       "      <td>6</td>\n",
       "      <td>PKC</td>\n",
       "      <td>DAG</td>\n",
       "      <td>[74, 77]</td>\n",
       "      <td>[16, 19]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, 2, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, 8, -1]</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12444</th>\n",
       "      <td>While SAR within the HTS series was very shallow and unable to be optimized, grafting the phenethyl ether linkage onto the ML129/ML172 cores led to the first sub-micromolar M5 PAM, ML326 (VU0467903), (human and rat M5 EC50s of 409nM and 500nM, respectively) with excellent mAChR selectivity (M1-M4 EC50s &gt;30μM) and a robust 20-fold leftward shift of the ACh CRC.</td>\n",
       "      <td>6</td>\n",
       "      <td>M5</td>\n",
       "      <td>phenethyl ether</td>\n",
       "      <td>[173, 175]</td>\n",
       "      <td>[90, 105]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>28.0</td>\n",
       "      <td>16.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3608</th>\n",
       "      <td>Ifenprodil is an allosteric inhibitor of GluN1/GluN2B N-methyl-D-aspartate receptors.</td>\n",
       "      <td>6</td>\n",
       "      <td>N-methyl-D-aspartate receptors</td>\n",
       "      <td>Ifenprodil</td>\n",
       "      <td>[54, 84]</td>\n",
       "      <td>[0, 10]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>7.5</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13508</th>\n",
       "      <td>Allosteric interaction of the neuromuscular blockers vecuronium and pancuronium with recombinant human muscarinic M2 receptors.</td>\n",
       "      <td>6</td>\n",
       "      <td>human muscarinic M2 receptors</td>\n",
       "      <td>vecuronium</td>\n",
       "      <td>[97, 126]</td>\n",
       "      <td>[53, 63]</td>\n",
       "      <td>[-1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>12.5</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13506</th>\n",
       "      <td>These data indicate that vecuronium, gallamine and pancuronium interact with an allosteric site on the muscarinic M2 receptor (located on the heart) and this may explain some of their cardiac side effects.</td>\n",
       "      <td>6</td>\n",
       "      <td>muscarinic M2 receptor</td>\n",
       "      <td>gallamine</td>\n",
       "      <td>[103, 125]</td>\n",
       "      <td>[37, 46]</td>\n",
       "      <td>[-1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>16.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11212</th>\n",
       "      <td>In addition to studies focused on the mechanisms of nuclear receptor function, additional work has illuminated the mechanism by which androgens are metabolized in selected tissues.</td>\n",
       "      <td>6</td>\n",
       "      <td>nuclear receptor</td>\n",
       "      <td>androgens</td>\n",
       "      <td>[52, 68]</td>\n",
       "      <td>[134, 143]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>9.5</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5377</th>\n",
       "      <td>We hypothesize that the rapamycin and related compounds bind to the α face and allosterically impact the proteasome function.</td>\n",
       "      <td>6</td>\n",
       "      <td>proteasome</td>\n",
       "      <td>rapamycin</td>\n",
       "      <td>[105, 115]</td>\n",
       "      <td>[24, 33]</td>\n",
       "      <td>[-1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>17.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11238</th>\n",
       "      <td>A number of agents are being developed that target molecular abnormalities in IEN, have fewer or different side effects than tamoxifen, and may be effective in ER-negative or tamoxifen-resistant disease.</td>\n",
       "      <td>6</td>\n",
       "      <td>ER</td>\n",
       "      <td>tamoxifen</td>\n",
       "      <td>[160, 162]</td>\n",
       "      <td>[125, 134]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>26.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5379</th>\n",
       "      <td>Rapamycin allosterically inhibits the proteasome.</td>\n",
       "      <td>6</td>\n",
       "      <td>proteasome</td>\n",
       "      <td>Rapamycin</td>\n",
       "      <td>[38, 48]</td>\n",
       "      <td>[0, 9]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12440</th>\n",
       "      <td>This Letter describes the further chemical optimization of the M5 PAM MLPCN probes ML129 and ML172.</td>\n",
       "      <td>6</td>\n",
       "      <td>M5</td>\n",
       "      <td>ML172</td>\n",
       "      <td>[63, 65]</td>\n",
       "      <td>[93, 98]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>9.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3478</th>\n",
       "      <td>Herein, we report the pharmacological actions of Lu AF21934, a novel, selective, and brain-penetrant positive allosteric modulator (PAM) of the mGlu(4) receptor in the stress-induced hyperthermia (SIH), four-plate, marble-burying and Vogel's conflict tests.</td>\n",
       "      <td>6</td>\n",
       "      <td>mGlu(4)</td>\n",
       "      <td>Lu AF21934</td>\n",
       "      <td>[144, 151]</td>\n",
       "      <td>[49, 59]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, 2, -1, -1, -1, -1, -1, -1, -1, -1, 6, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>20.0</td>\n",
       "      <td>7.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9859</th>\n",
       "      <td>It is concluded that salicylates are allosteric inhibitors of ETA receptors.</td>\n",
       "      <td>6</td>\n",
       "      <td>ETA receptors</td>\n",
       "      <td>salicylates</td>\n",
       "      <td>[62, 75]</td>\n",
       "      <td>[21, 32]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>9.5</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4409</th>\n",
       "      <td>However, one mGlu5 PAM, CPPHA (N-(4-chloro-2-[(1,3-dioxo-1,3-dihydro-2H-isoindol-2-yl)methyl]phenyl)-2-hydroxybenzamide), interacts with a separate allosteric site on mGlu5.</td>\n",
       "      <td>6</td>\n",
       "      <td>mGlu5</td>\n",
       "      <td>CPPHA</td>\n",
       "      <td>[13, 18]</td>\n",
       "      <td>[24, 29]</td>\n",
       "      <td>[-1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12441</th>\n",
       "      <td>Discovery of ML326: The first sub-micromolar, selective M5 PAM.</td>\n",
       "      <td>6</td>\n",
       "      <td>M5</td>\n",
       "      <td>ML326</td>\n",
       "      <td>[56, 58]</td>\n",
       "      <td>[13, 18]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3386</th>\n",
       "      <td>BDZs and other positive GABA(A)R modulators, including barbiturates, ethanol, and neurosteroids, can also inhibit L-type voltage-gated calcium channels (L-VGCCs), which could contribute to reduced neuronal excitability.</td>\n",
       "      <td>6</td>\n",
       "      <td>GABA(A)R</td>\n",
       "      <td>neurosteroids</td>\n",
       "      <td>[24, 32]</td>\n",
       "      <td>[82, 95]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, 3, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>4.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12437</th>\n",
       "      <td>While SAR within the HTS series was very shallow and unable to be optimized, grafting the phenethyl ether linkage onto the ML129/ML172 cores led to the first sub-micromolar M5 PAM, ML326 (VU0467903), (human and rat M5 EC50s of 409nM and 500nM, respectively) with excellent mAChR selectivity (M1-M4 EC50s &gt;30μM) and a robust 20-fold leftward shift of the ACh CRC.</td>\n",
       "      <td>6</td>\n",
       "      <td>M5</td>\n",
       "      <td>VU0467903</td>\n",
       "      <td>[173, 175]</td>\n",
       "      <td>[188, 197]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>28.0</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4410</th>\n",
       "      <td>However, one mGlu5 PAM, CPPHA (N-(4-chloro-2-[(1,3-dioxo-1,3-dihydro-2H-isoindol-2-yl)methyl]phenyl)-2-hydroxybenzamide), interacts with a separate allosteric site on mGlu5.</td>\n",
       "      <td>6</td>\n",
       "      <td>mGlu5</td>\n",
       "      <td>N-(4-chloro-2-[(1,3-dioxo-1,3-dihydro-2H-isoindol-2-yl)methyl]phenyl)-2-hydroxybenzamide</td>\n",
       "      <td>[13, 18]</td>\n",
       "      <td>[31, 119]</td>\n",
       "      <td>[-1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3383</th>\n",
       "      <td>Benzodiazepines (BDZs) depress neuronal excitability via positive allosteric modulation of inhibitory GABA(A) receptors (GABA(A)R).</td>\n",
       "      <td>6</td>\n",
       "      <td>GABA(A)R</td>\n",
       "      <td>Benzodiazepines</td>\n",
       "      <td>[121, 129]</td>\n",
       "      <td>[0, 15]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, 6, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14215</th>\n",
       "      <td>The structurally diverse opioids codeine and eseroline, like galantamine, are also nAChR-APL that have greatly diminished affinity for AChE, representing potential lead compounds for selective nAChR-APL development.</td>\n",
       "      <td>6</td>\n",
       "      <td>nAChR</td>\n",
       "      <td>eseroline</td>\n",
       "      <td>[83, 88]</td>\n",
       "      <td>[45, 54]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>11.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1136</th>\n",
       "      <td>A great deal of information has now accumulated pertaining to the mechanisms by which nuclear receptors, such as the androgen receptor, modulate the activity of responsive genes.</td>\n",
       "      <td>6</td>\n",
       "      <td>nuclear receptors</td>\n",
       "      <td>androgen</td>\n",
       "      <td>[86, 103]</td>\n",
       "      <td>[117, 125]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>14.5</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4413</th>\n",
       "      <td>Using cell-based assays and brain slice preparations, we characterized the interaction of a potent and efficacious mGlu5 PAM from the CPPHA series termed NCFP (N-(4-chloro-2-((4-fluoro-1,3-dioxoisoindolin-2-yl)methyl)phenyl)picolinamide).</td>\n",
       "      <td>6</td>\n",
       "      <td>mGlu5</td>\n",
       "      <td>N-(4-chloro-2-((4-fluoro-1,3-dioxoisoindolin-2-yl)methyl)phenyl)picolinamide</td>\n",
       "      <td>[115, 120]</td>\n",
       "      <td>[160, 236]</td>\n",
       "      <td>[-1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>16.0</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3385</th>\n",
       "      <td>BDZs and other positive GABA(A)R modulators, including barbiturates, ethanol, and neurosteroids, can also inhibit L-type voltage-gated calcium channels (L-VGCCs), which could contribute to reduced neuronal excitability.</td>\n",
       "      <td>6</td>\n",
       "      <td>GABA(A)R</td>\n",
       "      <td>ethanol</td>\n",
       "      <td>[24, 32]</td>\n",
       "      <td>[69, 76]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, 3, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3382</th>\n",
       "      <td>Benzodiazepines (BDZs) depress neuronal excitability via positive allosteric modulation of inhibitory GABA(A) receptors (GABA(A)R).</td>\n",
       "      <td>6</td>\n",
       "      <td>GABA(A) receptors</td>\n",
       "      <td>Benzodiazepines</td>\n",
       "      <td>[102, 119]</td>\n",
       "      <td>[0, 15]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, 6, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>11.5</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4411</th>\n",
       "      <td>Using cell-based assays and brain slice preparations, we characterized the interaction of a potent and efficacious mGlu5 PAM from the CPPHA series termed NCFP (N-(4-chloro-2-((4-fluoro-1,3-dioxoisoindolin-2-yl)methyl)phenyl)picolinamide).</td>\n",
       "      <td>6</td>\n",
       "      <td>mGlu5</td>\n",
       "      <td>CPPHA</td>\n",
       "      <td>[115, 120]</td>\n",
       "      <td>[134, 139]</td>\n",
       "      <td>[-1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>16.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12977</th>\n",
       "      <td>Diacylglycerol (DAG) acts as an allosteric activator of protein kinase C (PKC) and is converted to phosphatidic acid by DAG kinase (DGK).</td>\n",
       "      <td>6</td>\n",
       "      <td>protein kinase C</td>\n",
       "      <td>DAG</td>\n",
       "      <td>[56, 72]</td>\n",
       "      <td>[16, 19]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, 2, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, 8, -1]</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14214</th>\n",
       "      <td>The structurally diverse opioids codeine and eseroline, like galantamine, are also nAChR-APL that have greatly diminished affinity for AChE, representing potential lead compounds for selective nAChR-APL development.</td>\n",
       "      <td>6</td>\n",
       "      <td>nAChR</td>\n",
       "      <td>codeine</td>\n",
       "      <td>[193, 198]</td>\n",
       "      <td>[33, 40]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12976</th>\n",
       "      <td>Diacylglycerol (DAG) acts as an allosteric activator of protein kinase C (PKC) and is converted to phosphatidic acid by DAG kinase (DGK).</td>\n",
       "      <td>6</td>\n",
       "      <td>PKC</td>\n",
       "      <td>Diacylglycerol</td>\n",
       "      <td>[74, 77]</td>\n",
       "      <td>[0, 14]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, 2, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, 8, -1]</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8652</th>\n",
       "      <td>Conversely, ovarian PRA and PRB were positively regulated by ethanol and ethanol-melatonin combination, whereas PRA was down-regulated in the uterus and oviduct after ethanol consumption.</td>\n",
       "      <td>6</td>\n",
       "      <td>PRA</td>\n",
       "      <td>ethanol</td>\n",
       "      <td>[20, 23]</td>\n",
       "      <td>[73, 80]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 8, -1]</td>\n",
       "      <td>2.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4425</th>\n",
       "      <td>Evidence for a new binding mode to GSK-3: allosteric regulation by the marine compound palinurin.</td>\n",
       "      <td>6</td>\n",
       "      <td>GSK-3</td>\n",
       "      <td>palinurin</td>\n",
       "      <td>[35, 40]</td>\n",
       "      <td>[87, 96]</td>\n",
       "      <td>[-1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>7.0</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14213</th>\n",
       "      <td>The structurally diverse opioids codeine and eseroline, like galantamine, are also nAChR-APL that have greatly diminished affinity for AChE, representing potential lead compounds for selective nAChR-APL development.</td>\n",
       "      <td>6</td>\n",
       "      <td>nAChR</td>\n",
       "      <td>codeine</td>\n",
       "      <td>[83, 88]</td>\n",
       "      <td>[33, 40]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>11.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3763</th>\n",
       "      <td>Hydrogen sulfide as an allosteric modulator of ATP-sensitive potassium channels in colonic inflammation.</td>\n",
       "      <td>6</td>\n",
       "      <td>ATP-sensitive potassium channels</td>\n",
       "      <td>Hydrogen sulfide</td>\n",
       "      <td>[47, 79]</td>\n",
       "      <td>[0, 16]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3607</th>\n",
       "      <td>Ifenprodil is an allosteric inhibitor of GluN1/GluN2B N-methyl-D-aspartate receptors.</td>\n",
       "      <td>6</td>\n",
       "      <td>GluN2B</td>\n",
       "      <td>Ifenprodil</td>\n",
       "      <td>[47, 53]</td>\n",
       "      <td>[0, 10]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14212</th>\n",
       "      <td>Galantamine is a reversible, competitive acetylcholinesterase (AChE) inhibitor and allosteric potentiating ligand of nicotinic acetylcholine receptors (nAChR-APL) that shares many common structural elements with morphinan-based opioids.</td>\n",
       "      <td>6</td>\n",
       "      <td>nAChR</td>\n",
       "      <td>Galantamine</td>\n",
       "      <td>[152, 157]</td>\n",
       "      <td>[0, 11]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8653</th>\n",
       "      <td>Conversely, ovarian PRA and PRB were positively regulated by ethanol and ethanol-melatonin combination, whereas PRA was down-regulated in the uterus and oviduct after ethanol consumption.</td>\n",
       "      <td>6</td>\n",
       "      <td>PRB</td>\n",
       "      <td>ethanol</td>\n",
       "      <td>[28, 31]</td>\n",
       "      <td>[73, 80]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 8, -1]</td>\n",
       "      <td>4.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3384</th>\n",
       "      <td>BDZs and other positive GABA(A)R modulators, including barbiturates, ethanol, and neurosteroids, can also inhibit L-type voltage-gated calcium channels (L-VGCCs), which could contribute to reduced neuronal excitability.</td>\n",
       "      <td>6</td>\n",
       "      <td>GABA(A)R</td>\n",
       "      <td>barbiturates</td>\n",
       "      <td>[24, 32]</td>\n",
       "      <td>[55, 67]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, 3, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12443</th>\n",
       "      <td>An HTS campaign identified several weak M5 PAMs (M5 EC50 &gt;10μM) with a structurally related isatin core that possessed a southern phenethyl ether linkage.</td>\n",
       "      <td>6</td>\n",
       "      <td>M5</td>\n",
       "      <td>phenethyl ether</td>\n",
       "      <td>[49, 51]</td>\n",
       "      <td>[130, 145]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>8.0</td>\n",
       "      <td>21.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                                                                                                                                                                                                             text  \\\n",
       "12446  While SAR within the HTS series was very shallow and unable to be optimized, grafting the phenethyl ether linkage onto the ML129/ML172 cores led to the first sub-micromolar M5 PAM, ML326 (VU0467903), (human and rat M5 EC50s of 409nM and 500nM, respectively) with excellent mAChR selectivity (M1-M4 EC50s >30μM) and a robust 20-fold leftward shift of the ACh CRC.   \n",
       "3477                                                                                                                                                                                                                                      Anxiolytic- but not antidepressant-like activity of Lu AF21934, a novel, selective positive allosteric modulator of the mGlu₄ receptor.   \n",
       "8655                                                                                                                                                                                  Conversely, ovarian PRA and PRB were positively regulated by ethanol and ethanol-melatonin combination, whereas PRA was down-regulated in the uterus and oviduct after ethanol consumption.   \n",
       "3387                                                                                                                                                  BDZs and other positive GABA(A)R modulators, including barbiturates, ethanol, and neurosteroids, can also inhibit L-type voltage-gated calcium channels (L-VGCCs), which could contribute to reduced neuronal excitability.   \n",
       "8654                                                                                                                                                                                  Conversely, ovarian PRA and PRB were positively regulated by ethanol and ethanol-melatonin combination, whereas PRA was down-regulated in the uterus and oviduct after ethanol consumption.   \n",
       "13836                                                                                                                                                                                                          The recently discovered hyperinsulinism/hyperammonemia disorder showed that the loss of allosteric inhibition of GDH by GTP causes excessive secretion of insulin.   \n",
       "5378                                                                                                                                                                                                               The implications of our finding for mechanism of in vivo actions of rapamycin and for design of novel allosteric drugs targeting the proteasome are discussed.   \n",
       "3606                                                                                                                                                                                                                                                                                        Ifenprodil is an allosteric inhibitor of GluN1/GluN2B N-methyl-D-aspartate receptors.   \n",
       "15281                                                                                                                                                                                                                                       Synthesis and structure-activity relationships of indazole arylsulfonamides as allosteric CC-chemokine receptor 4 (CCR4) antagonists.   \n",
       "3381                                                                                                                                                                                                                                          Benzodiazepines (BDZs) depress neuronal excitability via positive allosteric modulation of inhibitory GABA(A) receptors (GABA(A)R).   \n",
       "14216                                                                                                                                                     The structurally diverse opioids codeine and eseroline, like galantamine, are also nAChR-APL that have greatly diminished affinity for AChE, representing potential lead compounds for selective nAChR-APL development.   \n",
       "8650                                                                                                                                                                                  Conversely, ovarian PRA and PRB were positively regulated by ethanol and ethanol-melatonin combination, whereas PRA was down-regulated in the uterus and oviduct after ethanol consumption.   \n",
       "12445  While SAR within the HTS series was very shallow and unable to be optimized, grafting the phenethyl ether linkage onto the ML129/ML172 cores led to the first sub-micromolar M5 PAM, ML326 (VU0467903), (human and rat M5 EC50s of 409nM and 500nM, respectively) with excellent mAChR selectivity (M1-M4 EC50s >30μM) and a robust 20-fold leftward shift of the ACh CRC.   \n",
       "9857                                                                                                                                                                                                                                                                           Aspirin and sodium salicylate inhibit endothelin ETA receptors by an allosteric type of mechanism.   \n",
       "13869                                                                                                               Because these speeds are significantly faster than the responses to antagonists, these data indicate that gallamine and dimethyl-W84 are allosteric ligands and actively induce a conformation of the M(2) receptor with a reduced affinity for its agonists.   \n",
       "8651                                                                                                                                                                                  Conversely, ovarian PRA and PRB were positively regulated by ethanol and ethanol-melatonin combination, whereas PRA was down-regulated in the uterus and oviduct after ethanol consumption.   \n",
       "5381                                                                                                                                                                                                                                                        Rapamycin is a canonical allosteric inhibitor of the mTOR kinase with immunosuppressive and pro-apoptotic activities.   \n",
       "12448  While SAR within the HTS series was very shallow and unable to be optimized, grafting the phenethyl ether linkage onto the ML129/ML172 cores led to the first sub-micromolar M5 PAM, ML326 (VU0467903), (human and rat M5 EC50s of 409nM and 500nM, respectively) with excellent mAChR selectivity (M1-M4 EC50s >30μM) and a robust 20-fold leftward shift of the ACh CRC.   \n",
       "12447  While SAR within the HTS series was very shallow and unable to be optimized, grafting the phenethyl ether linkage onto the ML129/ML172 cores led to the first sub-micromolar M5 PAM, ML326 (VU0467903), (human and rat M5 EC50s of 409nM and 500nM, respectively) with excellent mAChR selectivity (M1-M4 EC50s >30μM) and a robust 20-fold leftward shift of the ACh CRC.   \n",
       "12978                                                                                                                                                                                                                                   Diacylglycerol (DAG) acts as an allosteric activator of protein kinase C (PKC) and is converted to phosphatidic acid by DAG kinase (DGK).   \n",
       "12444  While SAR within the HTS series was very shallow and unable to be optimized, grafting the phenethyl ether linkage onto the ML129/ML172 cores led to the first sub-micromolar M5 PAM, ML326 (VU0467903), (human and rat M5 EC50s of 409nM and 500nM, respectively) with excellent mAChR selectivity (M1-M4 EC50s >30μM) and a robust 20-fold leftward shift of the ACh CRC.   \n",
       "3608                                                                                                                                                                                                                                                                                        Ifenprodil is an allosteric inhibitor of GluN1/GluN2B N-methyl-D-aspartate receptors.   \n",
       "13508                                                                                                                                                                                                                                             Allosteric interaction of the neuromuscular blockers vecuronium and pancuronium with recombinant human muscarinic M2 receptors.   \n",
       "13506                                                                                                                                                               These data indicate that vecuronium, gallamine and pancuronium interact with an allosteric site on the muscarinic M2 receptor (located on the heart) and this may explain some of their cardiac side effects.   \n",
       "11212                                                                                                                                                                                        In addition to studies focused on the mechanisms of nuclear receptor function, additional work has illuminated the mechanism by which androgens are metabolized in selected tissues.   \n",
       "5377                                                                                                                                                                                                                                                We hypothesize that the rapamycin and related compounds bind to the α face and allosterically impact the proteasome function.   \n",
       "11238                                                                                                                                                                 A number of agents are being developed that target molecular abnormalities in IEN, have fewer or different side effects than tamoxifen, and may be effective in ER-negative or tamoxifen-resistant disease.   \n",
       "5379                                                                                                                                                                                                                                                                                                                            Rapamycin allosterically inhibits the proteasome.   \n",
       "12440                                                                                                                                                                                                                                                                         This Letter describes the further chemical optimization of the M5 PAM MLPCN probes ML129 and ML172.   \n",
       "3478                                                                                                            Herein, we report the pharmacological actions of Lu AF21934, a novel, selective, and brain-penetrant positive allosteric modulator (PAM) of the mGlu(4) receptor in the stress-induced hyperthermia (SIH), four-plate, marble-burying and Vogel's conflict tests.   \n",
       "9859                                                                                                                                                                                                                                                                                                 It is concluded that salicylates are allosteric inhibitors of ETA receptors.   \n",
       "4409                                                                                                                                                                                                However, one mGlu5 PAM, CPPHA (N-(4-chloro-2-[(1,3-dioxo-1,3-dihydro-2H-isoindol-2-yl)methyl]phenyl)-2-hydroxybenzamide), interacts with a separate allosteric site on mGlu5.   \n",
       "12441                                                                                                                                                                                                                                                                                                             Discovery of ML326: The first sub-micromolar, selective M5 PAM.   \n",
       "3386                                                                                                                                                  BDZs and other positive GABA(A)R modulators, including barbiturates, ethanol, and neurosteroids, can also inhibit L-type voltage-gated calcium channels (L-VGCCs), which could contribute to reduced neuronal excitability.   \n",
       "12437  While SAR within the HTS series was very shallow and unable to be optimized, grafting the phenethyl ether linkage onto the ML129/ML172 cores led to the first sub-micromolar M5 PAM, ML326 (VU0467903), (human and rat M5 EC50s of 409nM and 500nM, respectively) with excellent mAChR selectivity (M1-M4 EC50s >30μM) and a robust 20-fold leftward shift of the ACh CRC.   \n",
       "4410                                                                                                                                                                                                However, one mGlu5 PAM, CPPHA (N-(4-chloro-2-[(1,3-dioxo-1,3-dihydro-2H-isoindol-2-yl)methyl]phenyl)-2-hydroxybenzamide), interacts with a separate allosteric site on mGlu5.   \n",
       "3383                                                                                                                                                                                                                                          Benzodiazepines (BDZs) depress neuronal excitability via positive allosteric modulation of inhibitory GABA(A) receptors (GABA(A)R).   \n",
       "14215                                                                                                                                                     The structurally diverse opioids codeine and eseroline, like galantamine, are also nAChR-APL that have greatly diminished affinity for AChE, representing potential lead compounds for selective nAChR-APL development.   \n",
       "1136                                                                                                                                                                                           A great deal of information has now accumulated pertaining to the mechanisms by which nuclear receptors, such as the androgen receptor, modulate the activity of responsive genes.   \n",
       "4413                                                                                                                               Using cell-based assays and brain slice preparations, we characterized the interaction of a potent and efficacious mGlu5 PAM from the CPPHA series termed NCFP (N-(4-chloro-2-((4-fluoro-1,3-dioxoisoindolin-2-yl)methyl)phenyl)picolinamide).   \n",
       "3385                                                                                                                                                  BDZs and other positive GABA(A)R modulators, including barbiturates, ethanol, and neurosteroids, can also inhibit L-type voltage-gated calcium channels (L-VGCCs), which could contribute to reduced neuronal excitability.   \n",
       "3382                                                                                                                                                                                                                                          Benzodiazepines (BDZs) depress neuronal excitability via positive allosteric modulation of inhibitory GABA(A) receptors (GABA(A)R).   \n",
       "4411                                                                                                                               Using cell-based assays and brain slice preparations, we characterized the interaction of a potent and efficacious mGlu5 PAM from the CPPHA series termed NCFP (N-(4-chloro-2-((4-fluoro-1,3-dioxoisoindolin-2-yl)methyl)phenyl)picolinamide).   \n",
       "12977                                                                                                                                                                                                                                   Diacylglycerol (DAG) acts as an allosteric activator of protein kinase C (PKC) and is converted to phosphatidic acid by DAG kinase (DGK).   \n",
       "14214                                                                                                                                                     The structurally diverse opioids codeine and eseroline, like galantamine, are also nAChR-APL that have greatly diminished affinity for AChE, representing potential lead compounds for selective nAChR-APL development.   \n",
       "12976                                                                                                                                                                                                                                   Diacylglycerol (DAG) acts as an allosteric activator of protein kinase C (PKC) and is converted to phosphatidic acid by DAG kinase (DGK).   \n",
       "8652                                                                                                                                                                                  Conversely, ovarian PRA and PRB were positively regulated by ethanol and ethanol-melatonin combination, whereas PRA was down-regulated in the uterus and oviduct after ethanol consumption.   \n",
       "4425                                                                                                                                                                                                                                                                            Evidence for a new binding mode to GSK-3: allosteric regulation by the marine compound palinurin.   \n",
       "14213                                                                                                                                                     The structurally diverse opioids codeine and eseroline, like galantamine, are also nAChR-APL that have greatly diminished affinity for AChE, representing potential lead compounds for selective nAChR-APL development.   \n",
       "3763                                                                                                                                                                                                                                                                     Hydrogen sulfide as an allosteric modulator of ATP-sensitive potassium channels in colonic inflammation.   \n",
       "3607                                                                                                                                                                                                                                                                                        Ifenprodil is an allosteric inhibitor of GluN1/GluN2B N-methyl-D-aspartate receptors.   \n",
       "14212                                                                                                                                Galantamine is a reversible, competitive acetylcholinesterase (AChE) inhibitor and allosteric potentiating ligand of nicotinic acetylcholine receptors (nAChR-APL) that shares many common structural elements with morphinan-based opioids.   \n",
       "8653                                                                                                                                                                                  Conversely, ovarian PRA and PRB were positively regulated by ethanol and ethanol-melatonin combination, whereas PRA was down-regulated in the uterus and oviduct after ethanol consumption.   \n",
       "3384                                                                                                                                                  BDZs and other positive GABA(A)R modulators, including barbiturates, ethanol, and neurosteroids, can also inhibit L-type voltage-gated calcium channels (L-VGCCs), which could contribute to reduced neuronal excitability.   \n",
       "12443                                                                                                                                                                                                                  An HTS campaign identified several weak M5 PAMs (M5 EC50 >10μM) with a structurally related isatin core that possessed a southern phenethyl ether linkage.   \n",
       "\n",
       "       labels                           entity1  \\\n",
       "12446       6                                M5   \n",
       "3477        6                             mGlu₄   \n",
       "8655        6                               PRB   \n",
       "3387        6                          GABA(A)R   \n",
       "8654        6                               PRA   \n",
       "13836       6                               GDH   \n",
       "5378        6                        proteasome   \n",
       "3606        6                             GluN1   \n",
       "15281       6           CC-chemokine receptor 4   \n",
       "3381        6                          GABA(A)R   \n",
       "14216       6                             nAChR   \n",
       "8650        6                               PRA   \n",
       "12445       6                                M5   \n",
       "9857        6                     ETA receptors   \n",
       "13869       6                     M(2) receptor   \n",
       "8651        6                               PRB   \n",
       "5381        6                            kinase   \n",
       "12448       6                                M5   \n",
       "12447       6                                M5   \n",
       "12978       6                               PKC   \n",
       "12444       6                                M5   \n",
       "3608        6    N-methyl-D-aspartate receptors   \n",
       "13508       6     human muscarinic M2 receptors   \n",
       "13506       6            muscarinic M2 receptor   \n",
       "11212       6                  nuclear receptor   \n",
       "5377        6                        proteasome   \n",
       "11238       6                                ER   \n",
       "5379        6                        proteasome   \n",
       "12440       6                                M5   \n",
       "3478        6                           mGlu(4)   \n",
       "9859        6                     ETA receptors   \n",
       "4409        6                             mGlu5   \n",
       "12441       6                                M5   \n",
       "3386        6                          GABA(A)R   \n",
       "12437       6                                M5   \n",
       "4410        6                             mGlu5   \n",
       "3383        6                          GABA(A)R   \n",
       "14215       6                             nAChR   \n",
       "1136        6                 nuclear receptors   \n",
       "4413        6                             mGlu5   \n",
       "3385        6                          GABA(A)R   \n",
       "3382        6                 GABA(A) receptors   \n",
       "4411        6                             mGlu5   \n",
       "12977       6                  protein kinase C   \n",
       "14214       6                             nAChR   \n",
       "12976       6                               PKC   \n",
       "8652        6                               PRA   \n",
       "4425        6                             GSK-3   \n",
       "14213       6                             nAChR   \n",
       "3763        6  ATP-sensitive potassium channels   \n",
       "3607        6                            GluN2B   \n",
       "14212       6                             nAChR   \n",
       "8653        6                               PRB   \n",
       "3384        6                          GABA(A)R   \n",
       "12443       6                                M5   \n",
       "\n",
       "                                                                                        entity2  \\\n",
       "12446                                                                                     ML172   \n",
       "3477                                                                                 Lu AF21934   \n",
       "8655                                                                                  melatonin   \n",
       "3387                                                                                       BDZs   \n",
       "8654                                                                                  melatonin   \n",
       "13836                                                                                       GTP   \n",
       "5378                                                                                  rapamycin   \n",
       "3606                                                                                 Ifenprodil   \n",
       "15281                                                                 indazole arylsulfonamides   \n",
       "3381                                                                                       BDZs   \n",
       "14216                                                                                 eseroline   \n",
       "8650                                                                                    ethanol   \n",
       "12445                                                                                     ML129   \n",
       "9857                                                                                    Aspirin   \n",
       "13869                                                                                 gallamine   \n",
       "8651                                                                                    ethanol   \n",
       "5381                                                                                  Rapamycin   \n",
       "12448                                                                                     ML326   \n",
       "12447                                                                                     ML326   \n",
       "12978                                                                                       DAG   \n",
       "12444                                                                           phenethyl ether   \n",
       "3608                                                                                 Ifenprodil   \n",
       "13508                                                                                vecuronium   \n",
       "13506                                                                                 gallamine   \n",
       "11212                                                                                 androgens   \n",
       "5377                                                                                  rapamycin   \n",
       "11238                                                                                 tamoxifen   \n",
       "5379                                                                                  Rapamycin   \n",
       "12440                                                                                     ML172   \n",
       "3478                                                                                 Lu AF21934   \n",
       "9859                                                                                salicylates   \n",
       "4409                                                                                      CPPHA   \n",
       "12441                                                                                     ML326   \n",
       "3386                                                                              neurosteroids   \n",
       "12437                                                                                 VU0467903   \n",
       "4410   N-(4-chloro-2-[(1,3-dioxo-1,3-dihydro-2H-isoindol-2-yl)methyl]phenyl)-2-hydroxybenzamide   \n",
       "3383                                                                            Benzodiazepines   \n",
       "14215                                                                                 eseroline   \n",
       "1136                                                                                   androgen   \n",
       "4413               N-(4-chloro-2-((4-fluoro-1,3-dioxoisoindolin-2-yl)methyl)phenyl)picolinamide   \n",
       "3385                                                                                    ethanol   \n",
       "3382                                                                            Benzodiazepines   \n",
       "4411                                                                                      CPPHA   \n",
       "12977                                                                                       DAG   \n",
       "14214                                                                                   codeine   \n",
       "12976                                                                            Diacylglycerol   \n",
       "8652                                                                                    ethanol   \n",
       "4425                                                                                  palinurin   \n",
       "14213                                                                                   codeine   \n",
       "3763                                                                           Hydrogen sulfide   \n",
       "3607                                                                                 Ifenprodil   \n",
       "14212                                                                               Galantamine   \n",
       "8653                                                                                    ethanol   \n",
       "3384                                                                               barbiturates   \n",
       "12443                                                                           phenethyl ether   \n",
       "\n",
       "            span1       span2  \\\n",
       "12446  [173, 175]  [129, 134]   \n",
       "3477   [120, 125]    [52, 62]   \n",
       "8655     [28, 31]    [81, 90]   \n",
       "3387     [24, 32]      [0, 4]   \n",
       "8654     [20, 23]    [81, 90]   \n",
       "13836  [113, 116]  [120, 123]   \n",
       "5378   [133, 143]    [68, 77]   \n",
       "3606     [41, 46]     [0, 10]   \n",
       "15281   [90, 113]    [50, 75]   \n",
       "3381   [121, 129]    [17, 21]   \n",
       "14216  [193, 198]    [45, 54]   \n",
       "8650     [20, 23]    [61, 68]   \n",
       "12445  [173, 175]  [123, 128]   \n",
       "9857     [49, 62]      [0, 7]   \n",
       "13869  [198, 211]  [106, 115]   \n",
       "8651     [28, 31]    [61, 68]   \n",
       "5381     [58, 64]      [0, 9]   \n",
       "12448  [215, 217]  [181, 186]   \n",
       "12447  [173, 175]  [181, 186]   \n",
       "12978    [74, 77]    [16, 19]   \n",
       "12444  [173, 175]   [90, 105]   \n",
       "3608     [54, 84]     [0, 10]   \n",
       "13508   [97, 126]    [53, 63]   \n",
       "13506  [103, 125]    [37, 46]   \n",
       "11212    [52, 68]  [134, 143]   \n",
       "5377   [105, 115]    [24, 33]   \n",
       "11238  [160, 162]  [125, 134]   \n",
       "5379     [38, 48]      [0, 9]   \n",
       "12440    [63, 65]    [93, 98]   \n",
       "3478   [144, 151]    [49, 59]   \n",
       "9859     [62, 75]    [21, 32]   \n",
       "4409     [13, 18]    [24, 29]   \n",
       "12441    [56, 58]    [13, 18]   \n",
       "3386     [24, 32]    [82, 95]   \n",
       "12437  [173, 175]  [188, 197]   \n",
       "4410     [13, 18]   [31, 119]   \n",
       "3383   [121, 129]     [0, 15]   \n",
       "14215    [83, 88]    [45, 54]   \n",
       "1136    [86, 103]  [117, 125]   \n",
       "4413   [115, 120]  [160, 236]   \n",
       "3385     [24, 32]    [69, 76]   \n",
       "3382   [102, 119]     [0, 15]   \n",
       "4411   [115, 120]  [134, 139]   \n",
       "12977    [56, 72]    [16, 19]   \n",
       "14214  [193, 198]    [33, 40]   \n",
       "12976    [74, 77]     [0, 14]   \n",
       "8652     [20, 23]    [73, 80]   \n",
       "4425     [35, 40]    [87, 96]   \n",
       "14213    [83, 88]    [33, 40]   \n",
       "3763     [47, 79]     [0, 16]   \n",
       "3607     [47, 53]     [0, 10]   \n",
       "14212  [152, 157]     [0, 11]   \n",
       "8653     [28, 31]    [73, 80]   \n",
       "3384     [24, 32]    [55, 67]   \n",
       "12443    [49, 51]  [130, 145]   \n",
       "\n",
       "                                                                                                    weak_labels  \\\n",
       "12446  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "3477      [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, 6, -1, -1, -1, -1, -1, -1, 9]   \n",
       "8655     [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 8, -1]   \n",
       "3387      [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, 3, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "8654     [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 8, -1]   \n",
       "13836    [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "5378    [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "3606     [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "15281    [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 5, -1, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "3381      [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, 6, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "14216   [-1, -1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "8650     [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 8, -1]   \n",
       "12445  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "9857     [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "13869        [-1, -1, -1, -1, -1, 1, -1, -1, 2, -1, -1, -1, 3, -1, -1, 4, 5, -1, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "8651     [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 8, -1]   \n",
       "5381     [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "12448  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "12447  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "12978     [-1, -1, -1, -1, -1, -1, 2, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, 8, -1]   \n",
       "12444  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "3608     [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "13508    [-1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "13506    [-1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "11212  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "5377     [-1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "11238  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "5379     [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "12440  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "3478      [-1, -1, -1, -1, -1, -1, -1, -1, 2, -1, -1, -1, -1, -1, -1, -1, -1, 6, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "9859     [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "4409     [-1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "12441  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "3386      [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, 3, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "12437  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "4410     [-1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "3383      [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, 6, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "14215   [-1, -1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "1136    [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "4413    [-1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "3385      [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, 3, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "3382      [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, 6, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "4411    [-1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "12977     [-1, -1, -1, -1, -1, -1, 2, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, 8, -1]   \n",
       "14214   [-1, -1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "12976     [-1, -1, -1, -1, -1, -1, 2, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, 8, -1]   \n",
       "8652     [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 8, -1]   \n",
       "4425     [-1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "14213   [-1, -1, -1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "3763     [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "3607     [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "14212    [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1]   \n",
       "8653     [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 8, -1]   \n",
       "3384      [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, 3, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "12443  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "\n",
       "       entity1_index  entity2_index  \n",
       "12446           28.0           21.0  \n",
       "3477            16.0            6.5  \n",
       "8655             4.0           11.0  \n",
       "3387             4.0            0.0  \n",
       "8654             2.0           11.0  \n",
       "13836           13.0           15.0  \n",
       "5378            22.0           12.0  \n",
       "3606             6.0            0.0  \n",
       "15281           10.0            5.5  \n",
       "3381            13.0            1.0  \n",
       "14216           25.0            6.0  \n",
       "8650             2.0            9.0  \n",
       "12445           28.0           21.0  \n",
       "9857             6.5            0.0  \n",
       "13869           28.5           15.0  \n",
       "8651             4.0            9.0  \n",
       "5381             9.0            0.0  \n",
       "12448           35.0           30.0  \n",
       "12447           28.0           30.0  \n",
       "12978           11.0            1.0  \n",
       "12444           28.0           16.5  \n",
       "3608             7.5            0.0  \n",
       "13508           12.5            6.0  \n",
       "13506           16.0            5.0  \n",
       "11212            9.5           20.0  \n",
       "5377            17.0            4.0  \n",
       "11238           26.0           20.0  \n",
       "5379             4.0            0.0  \n",
       "12440            9.0           15.0  \n",
       "3478            20.0            7.5  \n",
       "9859             9.5            4.0  \n",
       "4409             2.0            4.0  \n",
       "12441            7.0            2.0  \n",
       "3386             4.0           10.0  \n",
       "12437           28.0           31.0  \n",
       "4410             2.0            5.0  \n",
       "3383            13.0            0.0  \n",
       "14215           11.0            6.0  \n",
       "1136            14.5           19.0  \n",
       "4413            16.0           24.0  \n",
       "3385             4.0            8.0  \n",
       "3382            11.5            0.0  \n",
       "4411            16.0           20.0  \n",
       "12977            9.0            1.0  \n",
       "14214           25.0            4.0  \n",
       "12976           11.0            0.0  \n",
       "8652             2.0           11.0  \n",
       "4425             7.0           14.0  \n",
       "14213           11.0            4.0  \n",
       "3763             8.0            0.5  \n",
       "3607             6.0            0.0  \n",
       "14212           16.0            0.0  \n",
       "8653             4.0           11.0  \n",
       "3384             4.0            7.0  \n",
       "12443            8.0           21.5  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cell to inspect the dev set for a given label \n",
    "label = 'Modulator' # Change this variable\n",
    "pd.set_option('display.max_colwidth', 1000)\n",
    "label6 = train_df[train_df['labels'] == int(label_to_idx[label])]\n",
    "# save the dev set to a file\n",
    "df2 = li.chemprot_enhanced(label6)\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "3\n",
      "9\n",
      "29\n",
      "33\n",
      "36\n",
      "38\n",
      "40\n",
      "41\n",
      "49\n",
      "53\n",
      "count 11\n"
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "for i in range(len(df2)):\n",
    "    weak_label = df2.iloc[i]['weak_labels'][17]\n",
    "    if weak_label != -1:\n",
    "        print(i)\n",
    "        count+=1\n",
    "print('count', count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 55/55 [00:00<00:00, 22625.22it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[-1],\n",
       "       [ 6],\n",
       "       [-1],\n",
       "       [ 6],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [ 6],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [ 6],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [ 6],\n",
       "       [ 6],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [ 6],\n",
       "       [-1],\n",
       "       [ 6],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [ 6],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [ 6],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [ 6],\n",
       "       [ 6],\n",
       "       [-1],\n",
       "       [-1],\n",
       "       [ 6],\n",
       "       [-1]])"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "L_train = applier.apply(df=df2)\n",
    "L_train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>text</th>\n",
       "      <th>labels</th>\n",
       "      <th>entity1</th>\n",
       "      <th>entity2</th>\n",
       "      <th>span1</th>\n",
       "      <th>span2</th>\n",
       "      <th>weak_labels</th>\n",
       "      <th>entity1_index</th>\n",
       "      <th>entity2_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15523</th>\n",
       "      <td>This effect appeared to be due to both competition between S-nitrosocysteine and Prx1 for the Trx system and direct modulation by S-nitrosocysteine of Trx reductase activity.</td>\n",
       "      <td>1</td>\n",
       "      <td>Trx</td>\n",
       "      <td>S-nitrosocysteine</td>\n",
       "      <td>[94, 97]</td>\n",
       "      <td>[59, 76]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>15.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6532</th>\n",
       "      <td>Thalidomide--removed from widespread clinical use by 1962 because of severe teratogenicity--has anti-angiogenic and immunomodulatory effects, including the inhibition of TNF alpha.</td>\n",
       "      <td>3</td>\n",
       "      <td>TNF alpha</td>\n",
       "      <td>Thalidomide</td>\n",
       "      <td>[170, 179]</td>\n",
       "      <td>[0, 11]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>19.5</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5676</th>\n",
       "      <td>Data suggest that bisphosphonates via modulation of the activity of small-GTPases induce apoptosis in neoplastic cells by DNA-CpG-demethylation and stimulation of FAS-expression.</td>\n",
       "      <td>2</td>\n",
       "      <td>FAS</td>\n",
       "      <td>bisphosphonates</td>\n",
       "      <td>[163, 166]</td>\n",
       "      <td>[18, 33]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, 2, 2, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>21.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367</th>\n",
       "      <td>BDZs and other positive GABA(A)R modulators, including barbiturates, ethanol, and neurosteroids, can also inhibit L-type voltage-gated calcium channels (L-VGCCs), which could contribute to reduced neuronal excitability.</td>\n",
       "      <td>3</td>\n",
       "      <td>L-VGCCs</td>\n",
       "      <td>neurosteroids</td>\n",
       "      <td>[153, 160]</td>\n",
       "      <td>[82, 95]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, 3, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>18.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5005</th>\n",
       "      <td>This study is the first to identify Class V CGPs with their distinctive methine or trimethine linkage between two disubstituted pyrylium moieties as a particularly potent class of MRP modulators and also show that within this core structure, differences in the electronegativity associated with a chalcogen atom can be the sole determinant of whether a compound will stimulate or inhibit MRP2.</td>\n",
       "      <td>1</td>\n",
       "      <td>MRP</td>\n",
       "      <td>methine</td>\n",
       "      <td>[180, 183]</td>\n",
       "      <td>[72, 79]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, 2, -1, -1, -1, 3, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>28.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5116</th>\n",
       "      <td>Taken together, our findings indicate that 5HHMF suppresses NO production through modulation of iNOS, consequently suppressing NF-κB activity and induction of Nrf2-dependent HO-1 activity.</td>\n",
       "      <td>3</td>\n",
       "      <td>NF-κB</td>\n",
       "      <td>NO</td>\n",
       "      <td>[127, 132]</td>\n",
       "      <td>[60, 62]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, 2, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, 8, -1, -1]</td>\n",
       "      <td>16.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>491</th>\n",
       "      <td>There is, however, much information on the direct (acute and chronic) effects of alcohol on the binding properties of opioid receptors, as well as modulation of opioid peptide synthesis and secretion (e.g.</td>\n",
       "      <td>1</td>\n",
       "      <td>opioid peptide</td>\n",
       "      <td>alcohol</td>\n",
       "      <td>[161, 175]</td>\n",
       "      <td>[81, 88]</td>\n",
       "      <td>[-1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>26.5</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1225</th>\n",
       "      <td>Raloxifene is a selective ER modulator with less uterine estrogen agonist activity than tamoxifen, and it is hoped that it will result in fewer uterine cancers but will be equally efficacious in reducing the risk of breast cancer.</td>\n",
       "      <td>1</td>\n",
       "      <td>ER</td>\n",
       "      <td>Raloxifene</td>\n",
       "      <td>[26, 28]</td>\n",
       "      <td>[0, 10]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, 4, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3384</th>\n",
       "      <td>BDZs and other positive GABA(A)R modulators, including barbiturates, ethanol, and neurosteroids, can also inhibit L-type voltage-gated calcium channels (L-VGCCs), which could contribute to reduced neuronal excitability.</td>\n",
       "      <td>6</td>\n",
       "      <td>GABA(A)R</td>\n",
       "      <td>barbiturates</td>\n",
       "      <td>[24, 32]</td>\n",
       "      <td>[55, 67]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, 3, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7081</th>\n",
       "      <td>Similar to menthol, both camphor and cinnamaldehyde (initially reported to be specific activators of TRPV3 and TRPA1, respectively) also modulate other thermoTRPs.</td>\n",
       "      <td>2</td>\n",
       "      <td>TRPA1</td>\n",
       "      <td>menthol</td>\n",
       "      <td>[111, 116]</td>\n",
       "      <td>[11, 18]</td>\n",
       "      <td>[-1, -1, -1, -1, -1, -1, 2, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]</td>\n",
       "      <td>16.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>281 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                                                                                                                                                                                                                                            text  \\\n",
       "15523                                                                                                                                                                                                                             This effect appeared to be due to both competition between S-nitrosocysteine and Prx1 for the Trx system and direct modulation by S-nitrosocysteine of Trx reductase activity.   \n",
       "6532                                                                                                                                                                                                                        Thalidomide--removed from widespread clinical use by 1962 because of severe teratogenicity--has anti-angiogenic and immunomodulatory effects, including the inhibition of TNF alpha.   \n",
       "5676                                                                                                                                                                                                                          Data suggest that bisphosphonates via modulation of the activity of small-GTPases induce apoptosis in neoplastic cells by DNA-CpG-demethylation and stimulation of FAS-expression.   \n",
       "3367                                                                                                                                                                                 BDZs and other positive GABA(A)R modulators, including barbiturates, ethanol, and neurosteroids, can also inhibit L-type voltage-gated calcium channels (L-VGCCs), which could contribute to reduced neuronal excitability.   \n",
       "5005   This study is the first to identify Class V CGPs with their distinctive methine or trimethine linkage between two disubstituted pyrylium moieties as a particularly potent class of MRP modulators and also show that within this core structure, differences in the electronegativity associated with a chalcogen atom can be the sole determinant of whether a compound will stimulate or inhibit MRP2.   \n",
       "...                                                                                                                                                                                                                                                                                                                                                                                                          ...   \n",
       "5116                                                                                                                                                                                                                Taken together, our findings indicate that 5HHMF suppresses NO production through modulation of iNOS, consequently suppressing NF-κB activity and induction of Nrf2-dependent HO-1 activity.   \n",
       "491                                                                                                                                                                                                There is, however, much information on the direct (acute and chronic) effects of alcohol on the binding properties of opioid receptors, as well as modulation of opioid peptide synthesis and secretion (e.g.   \n",
       "1225                                                                                                                                                                      Raloxifene is a selective ER modulator with less uterine estrogen agonist activity than tamoxifen, and it is hoped that it will result in fewer uterine cancers but will be equally efficacious in reducing the risk of breast cancer.   \n",
       "3384                                                                                                                                                                                 BDZs and other positive GABA(A)R modulators, including barbiturates, ethanol, and neurosteroids, can also inhibit L-type voltage-gated calcium channels (L-VGCCs), which could contribute to reduced neuronal excitability.   \n",
       "7081                                                                                                                                                                                                                                         Similar to menthol, both camphor and cinnamaldehyde (initially reported to be specific activators of TRPV3 and TRPA1, respectively) also modulate other thermoTRPs.   \n",
       "\n",
       "       labels         entity1            entity2       span1     span2  \\\n",
       "15523       1             Trx  S-nitrosocysteine    [94, 97]  [59, 76]   \n",
       "6532        3       TNF alpha        Thalidomide  [170, 179]   [0, 11]   \n",
       "5676        2             FAS    bisphosphonates  [163, 166]  [18, 33]   \n",
       "3367        3         L-VGCCs      neurosteroids  [153, 160]  [82, 95]   \n",
       "5005        1             MRP            methine  [180, 183]  [72, 79]   \n",
       "...       ...             ...                ...         ...       ...   \n",
       "5116        3           NF-κB                 NO  [127, 132]  [60, 62]   \n",
       "491         1  opioid peptide            alcohol  [161, 175]  [81, 88]   \n",
       "1225        1              ER         Raloxifene    [26, 28]   [0, 10]   \n",
       "3384        6        GABA(A)R       barbiturates    [24, 32]  [55, 67]   \n",
       "7081        2           TRPA1            menthol  [111, 116]  [11, 18]   \n",
       "\n",
       "                                                                                                  weak_labels  \\\n",
       "15523  [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "6532   [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "5676    [-1, -1, -1, -1, -1, -1, -1, -1, 2, 2, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "3367    [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, 3, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "5005    [-1, -1, -1, -1, -1, -1, -1, -1, -1, 2, -1, -1, -1, 3, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "...                                                                                                       ...   \n",
       "5116    [-1, -1, -1, -1, -1, -1, -1, -1, 2, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, 8, -1, -1]   \n",
       "491    [-1, -1, -1, 1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "1225    [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, -1, -1, 4, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "3384    [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 3, 3, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "7081   [-1, -1, -1, -1, -1, -1, 2, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 6, -1, -1, -1, -1, -1, -1, -1, -1]   \n",
       "\n",
       "       entity1_index  entity2_index  \n",
       "15523           15.0           10.0  \n",
       "6532            19.5            0.0  \n",
       "5676            21.0            3.0  \n",
       "3367            18.0           10.0  \n",
       "5005            28.0           13.0  \n",
       "...              ...            ...  \n",
       "5116            16.0            8.0  \n",
       "491             26.5           13.0  \n",
       "1225             4.0            0.0  \n",
       "3384             4.0            7.0  \n",
       "7081            16.0            2.0  \n",
       "\n",
       "[281 rows x 9 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Cell to inspect the dev set for a given label \n",
    "label = 'Modulator' # Change this variable\n",
    "pd.set_option('display.max_colwidth', 1000)\n",
    "\n",
    "\n",
    "def get_18th_element(lst):\n",
    "    return lst[17] if len(lst) > 17  else None\n",
    "\n",
    "# Create a copy of the DataFrame\n",
    "train_df_copy = train_df.copy()\n",
    "\n",
    "# Apply the function to the copy\n",
    "train_df_copy['18th_weak_label'] = train_df_copy['weak_labels'].apply(get_18th_element)\n",
    "\n",
    "# Filter the copy\n",
    "label6_all = train_df[train_df_copy['18th_weak_label'] == 6]\n",
    "\n",
    "# Continue with your code\n",
    "df3 = li.chemprot_enhanced(label6_all)\n",
    "df3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n",
      "11\n",
      "count 2\n"
     ]
    }
   ],
   "source": [
    "df3 = df3.head(30)\n",
    "count = 0\n",
    "for i in range(len(df3)):\n",
    "    label = df3.iloc[i]['labels']\n",
    "    if label == 6:\n",
    "        print(i)\n",
    "        count+=1\n",
    "print('count', count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 30/30 [00:00<00:00, 23232.85it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n",
      "3\n",
      "5\n",
      "7\n",
      "11\n",
      "17\n",
      "20\n",
      "21\n",
      "26\n",
      "29\n",
      "count 10\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "L_train = applier.apply(df=df3)\n",
    "count = 0\n",
    "for i in range(len(L_train)):\n",
    "    if L_train[i][0] != -1:\n",
    "        print(i)\n",
    "        count+=1\n",
    "print('count', count)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Main Work"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2.to_csv(f\"./dev_{label}.csv\")\n",
    "chemprot6 = li.df_to_chemprot(li.chemprot_enhanced(label6))\n",
    "li.save_dataset(chemprot6, f\"./dev_{label}.json\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>text</th>\n",
       "      <th>labels</th>\n",
       "      <th>entity1</th>\n",
       "      <th>entity2</th>\n",
       "      <th>span1</th>\n",
       "      <th>span2</th>\n",
       "      <th>weak_labels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [text, labels, entity1, entity2, span1, span2, weak_labels]\n",
       "Index: []"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cell to inspect dev set for a given keyword (to inspect conflicts). Note: this doesn't work with \"+\" \n",
    "keyword = 'iPad' # Change this word\n",
    "pd.set_option('display.max_colwidth', 1000)\n",
    "dev_df[dev_df['text'].str.contains(keyword)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'apply_LFs' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_786926/3406333822.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mL_train\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mapply_LFs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlfs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtrain_df\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mL_valid\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mapply_LFs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlfs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalid_df\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mL_test\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mapply_LFs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlfs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtest_df\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Train Coverage:\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcalc_coverage\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mL_train\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'apply_LFs' is not defined"
     ]
    }
   ],
   "source": [
    "L_train = apply_LFs(lfs, train_df)\n",
    "L_valid = apply_LFs(lfs, valid_df)\n",
    "L_test = apply_LFs(lfs, test_df)\n",
    "\n",
    "print(\"Train Coverage:\", calc_coverage(L_train))\n",
    "print(\"Valid Coverage:\", calc_coverage(L_valid))\n",
    "print(\"Test Coverage:\", calc_coverage(L_test))\n",
    "\n",
    "lf_analysis = LFAnalysis(L=L_valid, lfs=lfs).lf_summary()\n",
    "\n",
    "# Calculates how many of an LFs votes result in conflicts (helpful signal for debugging LFs)\n",
    "lf_analysis['Conflict Ratio'] = lf_analysis['Conflicts'] / lf_analysis['Coverage']\n",
    "lf_analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# List LFs for which 'Conflict Ratio' is above some threshold (helpful for debugging)\n",
    "lf_analysis[lf_analysis['Conflict Ratio'] > 0.8]['Conflict Ratio'].sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'L_valid' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_989828/924070646.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Calculate accuracy on the validation set (Ideally do this only at the end)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mmajority_model\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mMajorityLabelVoter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mpreds_valid\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmajority_model\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpredict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mL\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mL_valid\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mpreds_valid\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mpreds_valid\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mvalid_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mpreds_valid\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmean\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'L_valid' is not defined"
     ]
    }
   ],
   "source": [
    "# Calculate accuracy on the validation set (Ideally do this only at the end)\n",
    "majority_model = MajorityLabelVoter(10)\n",
    "preds_valid = majority_model.predict(L)\n",
    "(preds_valid[preds_valid != -1] == valid_df[preds_valid != -1].labels.values).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "json.dump(keywords, open(\"amazon_LFs_v1.json\", \"w\"), indent=8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace the LFs for a given dataset (in wrench format)\n",
    "# dataset_name = \"dbpedia\"\n",
    "\n",
    "# train_json = json.load(open(f\"../weak_datasets/{dataset_name}/train.json\", \"r\"))\n",
    "# for idx in train_json:\n",
    "#     train_json[idx]['weak_labels'] = [int(i) for i in list(L_train[int(idx)])]\n",
    "    \n",
    "# valid_json = json.load(open(f\"../weak_datasets/{dataset_name}/valid.json\", \"r\"))\n",
    "# for idx in valid_json:\n",
    "#     valid_json[idx]['weak_labels'] = [int(i) for i in list(L_valid[int(idx)])]\n",
    "    \n",
    "# test_json = json.load(open(f\"../weak_datasets/{dataset_name}/test.json\", \"r\"))\n",
    "# for idx in test_json:\n",
    "#     test_json[idx]['weak_labels'] = [int(i) for i in list(L_test[int(idx)])]\n",
    "\n",
    "# json.dump(train_json, open(f\"../weak_datasets/{dataset_name}/train.json\", 'w'), indent=4)\n",
    "# json.dump(valid_json, open(f\"../weak_datasets/{dataset_name}/valid.json\", 'w'), indent=4)\n",
    "# json.dump(test_json, open(f\"../weak_datasets/{dataset_name}/test.json\", 'w'), indent=4)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  },
  "vscode": {
   "interpreter": {
    "hash": "3e41b164576941b46182852b40d7161f07e36b5af4c9af87265f406ae525ca4b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
